{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running scVI on donor P13 data:\n",
    "\n",
    "- SN (all) + multiome, all cell states\n",
    "\n",
    "Correcting by `sample`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "data_type = 'float32'\n",
    "os.environ[\"THEANO_FLAGS\"] = 'device=cuda,floatX=' + data_type + ',force_device=True'\n",
    "#sys.path.insert(1, '/nfs/team292/aa22/scVI_related/202105_troph_organoids/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Set up scVI environment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scvi\n",
    "import scanpy as sc\n",
    "\n",
    "sc.set_figure_params(figsize=(4, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Read in raw counts of the trophoblast organoids**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path = '/lustre/scratch117/cellgen/team292/aa22/adata_objects/202007_snRNA-seq_MFI/snRNA_seq_MFI_202007_adatas/202012_FINAL_reanalysis_with_souporcell_deconvolution_common_variants/donor_P13_all_samples_analysis_202110/'\n",
    "\n",
    "# Loading object with raw counts, filtered cells and genes object\n",
    "# before any doublet exclusion\n",
    "adata = sc.read(save_path + 'adata_raw_filtered.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>donor</th>\n",
       "      <th>tissue_block</th>\n",
       "      <th>age</th>\n",
       "      <th>sample</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>dataset</th>\n",
       "      <th>technique</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>...</th>\n",
       "      <th>barcode</th>\n",
       "      <th>souporcell_assignment</th>\n",
       "      <th>inter_ind_doublet</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>annotation_prev</th>\n",
       "      <th>louvain</th>\n",
       "      <th>is_doublet_propagate</th>\n",
       "      <th>origin_M_F</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGATACATG-1_WSSS_PLA8764121</th>\n",
       "      <td>1253</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1820.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.191617</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCAAGATACATG-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.053412</td>\n",
       "      <td>0.056947</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0_none</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA-1_WSSS_PLA8764121</th>\n",
       "      <td>1279</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>1868.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.032895</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAAATCCCA-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.120729</td>\n",
       "      <td>-0.113529</td>\n",
       "      <td>G1</td>\n",
       "      <td>dM2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACTGTGT-1_WSSS_PLA8764121</th>\n",
       "      <td>473</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>556.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.091892</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAACTGTGT-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.042173</td>\n",
       "      <td>-0.027934</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAGCTGCC-1_WSSS_PLA8764121</th>\n",
       "      <td>1360</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>2116.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.012335</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAAGCTGCC-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.121624</td>\n",
       "      <td>0.050396</td>\n",
       "      <td>G2M</td>\n",
       "      <td>dNK1</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACATAACCCA-1_WSSS_PLA8764121</th>\n",
       "      <td>1891</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>3274.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.073634</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACATAACCCA-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.125064</td>\n",
       "      <td>-0.066023</td>\n",
       "      <td>G1</td>\n",
       "      <td>dS1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>2409</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.007270</td>\n",
       "      <td>5227.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033632</td>\n",
       "      <td>-0.079893</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>4129</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>17379.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.102190</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.059688</td>\n",
       "      <td>-0.098701</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>3116</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.004499</td>\n",
       "      <td>7779.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.036876</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.066281</td>\n",
       "      <td>-0.038885</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>3696</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>20078.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.031627</td>\n",
       "      <td>-0.068024</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>2130</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.012930</td>\n",
       "      <td>5259.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.110820</td>\n",
       "      <td>-0.058170</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74244 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    n_genes donor  \\\n",
       "barcode_sample                                                      \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                     1253   P13   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                     1279   P13   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                      473   P13   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                     1360   P13   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                     1891   P13   \n",
       "...                                                     ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...     2409   P13   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...     4129   P13   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...     3116   P13   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...     3696   P13   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...     2130   P13   \n",
       "\n",
       "                                                   tissue_block      age  \\\n",
       "barcode_sample                                                             \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "...                                                         ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "\n",
       "                                                                                         sample  \\\n",
       "barcode_sample                                                                                    \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "...                                                                                         ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "\n",
       "                                                    percent_mito  n_counts  \\\n",
       "barcode_sample                                                               \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                      0.000000    1820.0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                      0.000535    1868.0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                      0.000000     556.0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                      0.000473    2116.0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                      0.000305    3274.0   \n",
       "...                                                          ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...      0.007270    5227.0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...      0.002532   17379.0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...      0.004499    7779.0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...      0.001992   20078.0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...      0.012930    5259.0   \n",
       "\n",
       "                                                      dataset technique  \\\n",
       "barcode_sample                                                            \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "...                                                       ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "\n",
       "                                                    scrublet_score  ...  \\\n",
       "barcode_sample                                                      ...   \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                        0.191617  ...   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                        0.032895  ...   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                        0.091892  ...   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                        0.012335  ...   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                        0.073634  ...   \n",
       "...                                                            ...  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...        0.013699  ...   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...        0.102190  ...   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...        0.036876  ...   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...        0.109137  ...   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...        0.109137  ...   \n",
       "\n",
       "                                                                                              barcode  \\\n",
       "barcode_sample                                                                                          \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                 AAACCCAAGATACATG-1_WSSS_PLA8764121   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                 AAACCCACAAATCCCA-1_WSSS_PLA8764121   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                 AAACCCACAACTGTGT-1_WSSS_PLA8764121   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                 AAACCCACAAGCTGCC-1_WSSS_PLA8764121   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                 AAACCCACATAACCCA-1_WSSS_PLA8764121   \n",
       "...                                                                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_...   \n",
       "\n",
       "                                                                            souporcell_assignment  \\\n",
       "barcode_sample                                                                                      \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                              WSSS_PLA8764121_0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "...                                                                                           ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "\n",
       "                                                    inter_ind_doublet  \\\n",
       "barcode_sample                                                          \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                  0   \n",
       "...                                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "\n",
       "                                                     S_score G2M_score phase  \\\n",
       "barcode_sample                                                                 \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                  0.053412  0.056947   G2M   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                 -0.120729 -0.113529    G1   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                  0.042173 -0.027934     S   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                 -0.121624  0.050396   G2M   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                 -0.125064 -0.066023    G1   \n",
       "...                                                      ...       ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  0.033632 -0.079893     S   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P... -0.059688 -0.098701    G1   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  0.066281 -0.038885     S   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P... -0.031627 -0.068024    G1   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P... -0.110820 -0.058170    G1   \n",
       "\n",
       "                                                   annotation_prev  louvain  \\\n",
       "barcode_sample                                                                \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                          0_none       16   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                             dM2       10   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                          0_none       14   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                            dNK1       14   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                             dS1        8   \n",
       "...                                                            ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...          0_none        7   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...          0_none       30   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "\n",
       "                                                    is_doublet_propagate  \\\n",
       "barcode_sample                                                             \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                     0   \n",
       "...                                                                  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "\n",
       "                                                    origin_M_F  \n",
       "barcode_sample                                                  \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                           F  \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                           M  \n",
       "...                                                        ...  \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...           M  \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "\n",
       "[74244 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading IDs of final cells to keep (after also excluding doublets by GEX, done in notebook M3)\n",
    "nodoublet_cell_IDs = pd.read_csv(save_path + 'obs_table_nodoublets.csv', index_col=0)\n",
    "nodoublet_cell_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74244 × 29058\n",
       "    obs: 'n_genes', 'donor', 'tissue_block', 'age', 'sample', 'percent_mito', 'n_counts', 'dataset', 'technique', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch', 'is_doublet'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subsetting the adata object to final cells\n",
    "adata = adata[list(nodoublet_cell_IDs.index), :].copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_genes\n",
      "donor\n",
      "tissue_block\n",
      "age\n",
      "sample\n",
      "percent_mito\n",
      "n_counts\n",
      "dataset\n",
      "technique\n",
      "scrublet_score\n",
      "scrublet_cluster_score\n",
      "bh_pval\n",
      "batch\n",
      "is_doublet\n",
      "barcode_sample_copy\n",
      "barcode\n",
      "souporcell_assignment\n",
      "inter_ind_doublet\n",
      "S_score\n",
      "G2M_score\n",
      "phase\n",
      "annotation_prev\n",
      "louvain\n",
      "is_doublet_propagate\n",
      "origin_M_F\n"
     ]
    }
   ],
   "source": [
    "# adding metadata\n",
    "for col in nodoublet_cell_IDs.columns:\n",
    "    print(col)\n",
    "    adata.obs[col] = nodoublet_cell_IDs.loc[adata.obs_names,col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>feature_types-0</th>\n",
       "      <th>genome-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>gene_ids-1</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>genome-1</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>gene_ids-2</th>\n",
       "      <th>feature_types-2</th>\n",
       "      <th>...</th>\n",
       "      <th>genome-5</th>\n",
       "      <th>n_cells-5</th>\n",
       "      <th>gene_ids-6</th>\n",
       "      <th>feature_types-6</th>\n",
       "      <th>genome-6</th>\n",
       "      <th>n_cells-6</th>\n",
       "      <th>gene_ids-7</th>\n",
       "      <th>feature_types-7</th>\n",
       "      <th>genome-7</th>\n",
       "      <th>n_cells-7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>105.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>119.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>101.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>159.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>160.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>162.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>211.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>152.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>223.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>217.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>12.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>11.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1302.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1356.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1074.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1429.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1333.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>20.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>18.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>10.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>36.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1994.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>2274.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>2410.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2541.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>218.0</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>276.0</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>229.0</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>216.0</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>226.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>2320.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>2526.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>2920.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2735.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2810.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1274.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1402.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1795.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1690.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>3292.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>3535.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "      <td>4151.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>3701.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>3752.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28854 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_ids-0  feature_types-0              genome-0  n_cells-0  \\\n",
       "A1BG      ENSG00000121410  Gene Expression  GRCh38-3.0.0_premrna      105.0   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression  GRCh38-3.0.0_premrna      162.0   \n",
       "A1CF      ENSG00000148584  Gene Expression  GRCh38-3.0.0_premrna       12.0   \n",
       "A2M       ENSG00000175899  Gene Expression  GRCh38-3.0.0_premrna     1302.0   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression  GRCh38-3.0.0_premrna       20.0   \n",
       "...                   ...              ...                   ...        ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression  GRCh38-3.0.0_premrna     1994.0   \n",
       "ZYG11A    ENSG00000203995  Gene Expression  GRCh38-3.0.0_premrna      218.0   \n",
       "ZYG11B    ENSG00000162378  Gene Expression  GRCh38-3.0.0_premrna     2320.0   \n",
       "ZYX       ENSG00000159840  Gene Expression  GRCh38-3.0.0_premrna     1274.0   \n",
       "ZZEF1     ENSG00000074755  Gene Expression  GRCh38-3.0.0_premrna     3292.0   \n",
       "\n",
       "               gene_ids-1  feature_types-1              genome-1  n_cells-1  \\\n",
       "A1BG      ENSG00000121410  Gene Expression  GRCh38-3.0.0_premrna      119.0   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression  GRCh38-3.0.0_premrna      211.0   \n",
       "A1CF      ENSG00000148584  Gene Expression  GRCh38-3.0.0_premrna       11.0   \n",
       "A2M       ENSG00000175899  Gene Expression  GRCh38-3.0.0_premrna     1356.0   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression  GRCh38-3.0.0_premrna       18.0   \n",
       "...                   ...              ...                   ...        ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression  GRCh38-3.0.0_premrna     2274.0   \n",
       "ZYG11A    ENSG00000203995  Gene Expression  GRCh38-3.0.0_premrna      276.0   \n",
       "ZYG11B    ENSG00000162378  Gene Expression  GRCh38-3.0.0_premrna     2526.0   \n",
       "ZYX       ENSG00000159840  Gene Expression  GRCh38-3.0.0_premrna     1402.0   \n",
       "ZZEF1     ENSG00000074755  Gene Expression  GRCh38-3.0.0_premrna     3535.0   \n",
       "\n",
       "               gene_ids-2  feature_types-2  ...              genome-5  \\\n",
       "A1BG      ENSG00000121410  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "A1CF      ENSG00000148584  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "A2M       ENSG00000175899  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "...                   ...              ...  ...                   ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "ZYG11A    ENSG00000203995  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "ZYG11B    ENSG00000162378  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "ZYX       ENSG00000159840  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "ZZEF1     ENSG00000074755  Gene Expression  ...  GRCh38-3.0.0_premrna   \n",
       "\n",
       "          n_cells-5       gene_ids-6  feature_types-6 genome-6  n_cells-6  \\\n",
       "A1BG          101.0  ENSG00000121410  Gene Expression   GRCh38      159.0   \n",
       "A1BG-AS1      152.0  ENSG00000268895  Gene Expression   GRCh38      223.0   \n",
       "A1CF            9.0  ENSG00000148584  Gene Expression   GRCh38       12.0   \n",
       "A2M          1074.0  ENSG00000175899  Gene Expression   GRCh38     1429.0   \n",
       "A2M-AS1        10.0  ENSG00000245105  Gene Expression   GRCh38       36.0   \n",
       "...             ...              ...              ...      ...        ...   \n",
       "ZXDC         2410.0  ENSG00000070476  Gene Expression   GRCh38     2541.0   \n",
       "ZYG11A        229.0  ENSG00000203995  Gene Expression   GRCh38      216.0   \n",
       "ZYG11B       2920.0  ENSG00000162378  Gene Expression   GRCh38     2735.0   \n",
       "ZYX          1124.0  ENSG00000159840  Gene Expression   GRCh38     1795.0   \n",
       "ZZEF1        4151.0  ENSG00000074755  Gene Expression   GRCh38     3701.0   \n",
       "\n",
       "               gene_ids-7  feature_types-7 genome-7  n_cells-7  \n",
       "A1BG      ENSG00000121410  Gene Expression   GRCh38      160.0  \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression   GRCh38      217.0  \n",
       "A1CF      ENSG00000148584  Gene Expression   GRCh38       15.0  \n",
       "A2M       ENSG00000175899  Gene Expression   GRCh38     1333.0  \n",
       "A2M-AS1   ENSG00000245105  Gene Expression   GRCh38       25.0  \n",
       "...                   ...              ...      ...        ...  \n",
       "ZXDC      ENSG00000070476  Gene Expression   GRCh38     2641.0  \n",
       "ZYG11A    ENSG00000203995  Gene Expression   GRCh38      226.0  \n",
       "ZYG11B    ENSG00000162378  Gene Expression   GRCh38     2810.0  \n",
       "ZYX       ENSG00000159840  Gene Expression   GRCh38     1690.0  \n",
       "ZZEF1     ENSG00000074755  Gene Expression   GRCh38     3752.0  \n",
       "\n",
       "[28854 rows x 32 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# addition - subsetting to genes after cell cycle-associated gene removal\n",
    "# read in cleaned up genes, after all doublet exclusion\n",
    "# this is done in notebook S3\n",
    "cleaned_up_genes = pd.read_csv(save_path + 'genes_without_CC_assoc_genes_for_scVI_20211109.csv', \n",
    "                              index_col=0)\n",
    "cleaned_up_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subsetting to only cleamed up genes\n",
    "adata = adata[:, list(cleaned_up_genes.index)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74244 × 28854\n",
       "    obs: 'n_genes', 'donor', 'tissue_block', 'age', 'sample', 'percent_mito', 'n_counts', 'dataset', 'technique', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch', 'is_doublet', 'barcode_sample_copy', 'barcode', 'souporcell_assignment', 'inter_ind_doublet', 'S_score', 'G2M_score', 'phase', 'annotation_prev', 'louvain', 'is_doublet_propagate', 'origin_M_F'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# also getting rid of soupy genes \n",
    "# optional, trying without for now\n",
    "#gene_stats = pd.read_csv('./gene_filter_cells_by_sample_Pla_Camb10123930_and_Pla_Camb10123931.csv', index_col=0)\n",
    "#gene_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#np.unique(gene_stats['Selected'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes2keep = list(gene_stats[gene_stats['Selected'] == True].index)\n",
    "#genes2keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subsetting to only cleaned up genes\n",
    "#adata = adata[:, list(set(genes2keep) & set(adata.var_names))].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74244 × 28854\n",
       "    obs: 'n_genes', 'donor', 'tissue_block', 'age', 'sample', 'percent_mito', 'n_counts', 'dataset', 'technique', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch', 'is_doublet', 'barcode_sample_copy', 'barcode', 'souporcell_assignment', 'inter_ind_doublet', 'S_score', 'G2M_score', 'phase', 'annotation_prev', 'louvain', 'is_doublet_propagate', 'origin_M_F'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['Pla_Camb10691970', 'Pla_Camb10691971',\n",
       "        'Pla_Camb10714919_and_40110_Pla_Camb10687915',\n",
       "        'Pla_Camb10714920_and_40110_Pla_Camb10687916', 'WSSS_PLA8764121',\n",
       "        'WSSS_PLA8764122', 'WSSS_PLA8810750', 'WSSS_PLA8810751'],\n",
       "       dtype=object),\n",
       " array([ 9740,  9840,  5172,  5407,  8658,  8411, 14455, 12561]))"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# going to correct by sample + donor here\n",
    "np.unique(adata.obs['sample'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do some basic preprocessing\n",
    "adata.layers[\"raw_counts\"] = adata.X.copy() # preserve counts\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "adata.raw = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating a joint donor + sample key for the most correct HVG calculation in case of donor + sample correction\n",
    "#adata.obs['donor_sample'] = [i + '_' + j for i,j in zip(adata.obs['donor'], adata.obs['sample'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'donor' as categorical\n",
      "... storing 'tissue_block' as categorical\n",
      "... storing 'age' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'technique' as categorical\n",
      "... storing 'souporcell_assignment' as categorical\n",
      "... storing 'phase' as categorical\n",
      "... storing 'annotation_prev' as categorical\n",
      "... storing 'origin_M_F' as categorical\n",
      "/opt/conda/envs/scvi-singularity/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    }
   ],
   "source": [
    "# here we subset to HVGs, accounting for donor batch\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata,\n",
    "    n_top_genes=2000,\n",
    "    subset=True,\n",
    "    #layer=\"raw_counts\",\n",
    "    flavor=\"seurat\",\n",
    "    batch_key=\"sample\",\n",
    "    #batch_key=\"donor\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74244 × 2000\n",
       "    obs: 'n_genes', 'donor', 'tissue_block', 'age', 'sample', 'percent_mito', 'n_counts', 'dataset', 'technique', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch', 'is_doublet', 'barcode_sample_copy', 'barcode', 'souporcell_assignment', 'inter_ind_doublet', 'S_score', 'G2M_score', 'phase', 'annotation_prev', 'louvain', 'is_doublet_propagate', 'origin_M_F'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'log1p', 'hvg'\n",
       "    layers: 'raw_counts'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here you can specify your covariates you want to correct for (categorical or continuous)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m No batch_key inputted, assuming all cells are same batch                            \n",
      "\u001b[34mINFO    \u001b[0m No label_key inputted, assuming all cells have same label                           \n",
      "\u001b[34mINFO    \u001b[0m Using data from adata.layers\u001b[1m[\u001b[0m\u001b[32m\"raw_counts\"\u001b[0m\u001b[1m]\u001b[0m                                          \n",
      "\u001b[34mINFO    \u001b[0m Computing library size prior per batch                                              \n",
      "\u001b[34mINFO    \u001b[0m Successfully registered anndata object containing \u001b[1;36m74244\u001b[0m cells, \u001b[1;36m2000\u001b[0m vars, \u001b[1;36m1\u001b[0m batches,\n",
      "         \u001b[1;36m1\u001b[0m labels, and \u001b[1;36m0\u001b[0m proteins. Also registered \u001b[1;36m1\u001b[0m extra categorical covariates and \u001b[1;36m0\u001b[0m extra\n",
      "         continuous covariates.                                                              \n",
      "\u001b[34mINFO    \u001b[0m Please do not further modify adata until model is trained.                          \n"
     ]
    }
   ],
   "source": [
    "scvi.data.setup_anndata(\n",
    "    adata,\n",
    "    layer=\"raw_counts\",\n",
    "    categorical_covariate_keys=[\"sample\"],\n",
    "    #continuous_covariate_keys=[\"\"] # could try and regress n_genes in the future? not for now\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 09.11.2021 Running scVI after:\n",
    "\n",
    "- removing cell cycle-associated genes (calculated in S3 notebook)\n",
    "- exclusion of all doublets (scrublet and souporcell, done in S2 notebook)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_latent_value 5\n",
      "n_latent_value 10\n",
      "n_latent_value 15\n",
      "n_latent_value 17\n",
      "n_latent_value 20\n",
      "n_latent_value 30\n",
      "n_latent_value 40\n"
     ]
    }
   ],
   "source": [
    "models = {}\n",
    "\n",
    "# just a few values here, it was 17 PCs I used in the conventional analysis as optimal n_PCs\n",
    "n_latent_values = [5, 10, 15, \n",
    "                   17,\n",
    "                   20, 30, 40]\n",
    "\n",
    "for n_latent_value in n_latent_values:\n",
    "    print('n_latent_value', n_latent_value)\n",
    "    models[n_latent_value] = scvi.model.SCVI(adata, n_latent = n_latent_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCVI Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">128</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">40</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.1</span>, dispersion: gene, \n",
       "gene_likelihood: zinb, latent_distribution: normal\n",
       "Training status: Not Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCVI Model with the following params: \n",
       "n_hidden: \u001b[1;36m128\u001b[0m, n_latent: \u001b[1;36m40\u001b[0m, n_layers: \u001b[1;36m1\u001b[0m, dropout_rate: \u001b[1;36m0.1\u001b[0m, dispersion: gene, \n",
       "gene_likelihood: zinb, latent_distribution: normal\n",
       "Training status: Not Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "\n",
       "To print summary of associated AnnData, use: <span style=\"color: #af00d7; text-decoration-color: #af00d7\">scvi.data.view_anndata_setup(model.adata)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n",
       "\n",
       "To print summary of associated AnnData, use: \u001b[38;5;128mscvi.data.view_anndata_setup(model.adata)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models[40]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 5\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:49<00:00,  4.90s/it, loss=589, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 10\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:49<00:00,  4.90s/it, loss=603, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 15\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:51<00:00,  4.92s/it, loss=590, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 20\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:50<00:00,  4.91s/it, loss=601, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 30\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:51<00:00,  4.92s/it, loss=583, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 40\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:51<00:00,  4.92s/it, loss=592, v_num=1]\n"
     ]
    }
   ],
   "source": [
    "latent_representations = {}\n",
    "\n",
    "for n_latent_value in n_latent_values:\n",
    "    print('training model for n_latent_value:', n_latent_value)\n",
    "    models[n_latent_value].train()\n",
    "    \n",
    "    # get latent representation as a dataframe\n",
    "    latent_representations[n_latent_value] = models[n_latent_value].get_latent_representation()\n",
    "    \n",
    "    # add it to the adata object\n",
    "    adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)] = latent_representations[n_latent_value]\n",
    "    curr_df = pd.DataFrame(adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)])\n",
    "    \n",
    "    # save the latent representation\n",
    "    curr_df.to_csv('./results/20211109_obsm_with_scVI_latent_representation_n_' + str(n_latent_value) + '_after_CCG_removal_and_all_doublet_exclusion_by_sample_donor_P13_all_SN_and_multiome.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 17\n",
      "Epoch 108/108: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108/108 [08:50<00:00,  4.91s/it, loss=592, v_num=1]\n"
     ]
    }
   ],
   "source": [
    "# calculating with n_latent = 17 since 17 PCs best explained the variance in the object in conventional analysis\n",
    "\n",
    "latent_representations = {}\n",
    "\n",
    "for n_latent_value in [17]:\n",
    "    print('training model for n_latent_value:', n_latent_value)\n",
    "    models[n_latent_value].train()\n",
    "    \n",
    "    # get latent representation as a dataframe\n",
    "    latent_representations[n_latent_value] = models[n_latent_value].get_latent_representation()\n",
    "    \n",
    "    # add it to the adata object\n",
    "    adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)] = latent_representations[n_latent_value]\n",
    "    curr_df = pd.DataFrame(adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)])\n",
    "    \n",
    "    # save the latent representation\n",
    "    curr_df.to_csv('./20211109_obsm_with_scVI_latent_representation_n_' + str(n_latent_value) + '_after_CCG_removal_and_all_doublet_exclusion_by_sample_donor_P13_all_SN_and_multiome.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# after this, going to notebook S4 to calculate a manifold and have a look at it etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code below is old - pls ignore"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running on only invading trophoblast \n",
    "\n",
    "\n",
    "Subsetted by unbiased clusteing of the manifold with all cell states as calculated above with n_latent=17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path = '/lustre/scratch117/cellgen/team292/aa22/adata_objects/202007_snRNA-seq_MFI/snRNA_seq_MFI_202007_adatas/202012_FINAL_reanalysis_with_souporcell_deconvolution_common_variants/donor_P13_all_samples_analysis_202110/'\n",
    "\n",
    "# Loading object with raw counts, filtered cells and genes object\n",
    "# before any doublet exclusion\n",
    "adata = sc.read(save_path + 'adata_raw_filtered.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>donor</th>\n",
       "      <th>tissue_block</th>\n",
       "      <th>age</th>\n",
       "      <th>sample</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>dataset</th>\n",
       "      <th>technique</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>...</th>\n",
       "      <th>barcode</th>\n",
       "      <th>souporcell_assignment</th>\n",
       "      <th>inter_ind_doublet</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>annotation_prev</th>\n",
       "      <th>louvain</th>\n",
       "      <th>is_doublet_propagate</th>\n",
       "      <th>origin_M_F</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGATACATG-1_WSSS_PLA8764121</th>\n",
       "      <td>1253</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1820.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.191617</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCAAGATACATG-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.053412</td>\n",
       "      <td>0.056947</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0_none</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA-1_WSSS_PLA8764121</th>\n",
       "      <td>1279</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>1868.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.032895</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAAATCCCA-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.120729</td>\n",
       "      <td>-0.113529</td>\n",
       "      <td>G1</td>\n",
       "      <td>dM2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACTGTGT-1_WSSS_PLA8764121</th>\n",
       "      <td>473</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>556.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.091892</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAACTGTGT-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.042173</td>\n",
       "      <td>-0.027934</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAGCTGCC-1_WSSS_PLA8764121</th>\n",
       "      <td>1360</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>2116.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.012335</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAAGCTGCC-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.121624</td>\n",
       "      <td>0.050396</td>\n",
       "      <td>G2M</td>\n",
       "      <td>dNK1</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACATAACCCA-1_WSSS_PLA8764121</th>\n",
       "      <td>1891</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>3274.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.073634</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACATAACCCA-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.125064</td>\n",
       "      <td>-0.066023</td>\n",
       "      <td>G1</td>\n",
       "      <td>dS1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>2409</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.007270</td>\n",
       "      <td>5227.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033632</td>\n",
       "      <td>-0.079893</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>4129</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>17379.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.102190</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.059688</td>\n",
       "      <td>-0.098701</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>3116</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.004499</td>\n",
       "      <td>7779.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.036876</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.066281</td>\n",
       "      <td>-0.038885</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>3696</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>20078.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.031627</td>\n",
       "      <td>-0.068024</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>2130</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.012930</td>\n",
       "      <td>5259.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.110820</td>\n",
       "      <td>-0.058170</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74244 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    n_genes donor  \\\n",
       "barcode_sample                                                      \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                     1253   P13   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                     1279   P13   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                      473   P13   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                     1360   P13   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                     1891   P13   \n",
       "...                                                     ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...     2409   P13   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...     4129   P13   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...     3116   P13   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...     3696   P13   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...     2130   P13   \n",
       "\n",
       "                                                   tissue_block      age  \\\n",
       "barcode_sample                                                             \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "...                                                         ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "\n",
       "                                                                                         sample  \\\n",
       "barcode_sample                                                                                    \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "...                                                                                         ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "\n",
       "                                                    percent_mito  n_counts  \\\n",
       "barcode_sample                                                               \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                      0.000000    1820.0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                      0.000535    1868.0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                      0.000000     556.0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                      0.000473    2116.0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                      0.000305    3274.0   \n",
       "...                                                          ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...      0.007270    5227.0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...      0.002532   17379.0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...      0.004499    7779.0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...      0.001992   20078.0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...      0.012930    5259.0   \n",
       "\n",
       "                                                      dataset technique  \\\n",
       "barcode_sample                                                            \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "...                                                       ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "\n",
       "                                                    scrublet_score  ...  \\\n",
       "barcode_sample                                                      ...   \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                        0.191617  ...   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                        0.032895  ...   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                        0.091892  ...   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                        0.012335  ...   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                        0.073634  ...   \n",
       "...                                                            ...  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...        0.013699  ...   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...        0.102190  ...   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...        0.036876  ...   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...        0.109137  ...   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...        0.109137  ...   \n",
       "\n",
       "                                                                                              barcode  \\\n",
       "barcode_sample                                                                                          \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                 AAACCCAAGATACATG-1_WSSS_PLA8764121   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                 AAACCCACAAATCCCA-1_WSSS_PLA8764121   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                 AAACCCACAACTGTGT-1_WSSS_PLA8764121   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                 AAACCCACAAGCTGCC-1_WSSS_PLA8764121   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                 AAACCCACATAACCCA-1_WSSS_PLA8764121   \n",
       "...                                                                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_...   \n",
       "\n",
       "                                                                            souporcell_assignment  \\\n",
       "barcode_sample                                                                                      \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                              WSSS_PLA8764121_0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "...                                                                                           ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "\n",
       "                                                    inter_ind_doublet  \\\n",
       "barcode_sample                                                          \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                  0   \n",
       "...                                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "\n",
       "                                                     S_score G2M_score phase  \\\n",
       "barcode_sample                                                                 \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                  0.053412  0.056947   G2M   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                 -0.120729 -0.113529    G1   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                  0.042173 -0.027934     S   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                 -0.121624  0.050396   G2M   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                 -0.125064 -0.066023    G1   \n",
       "...                                                      ...       ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  0.033632 -0.079893     S   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P... -0.059688 -0.098701    G1   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  0.066281 -0.038885     S   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P... -0.031627 -0.068024    G1   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P... -0.110820 -0.058170    G1   \n",
       "\n",
       "                                                   annotation_prev  louvain  \\\n",
       "barcode_sample                                                                \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                          0_none       16   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                             dM2       10   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                          0_none       14   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                            dNK1       14   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                             dS1        8   \n",
       "...                                                            ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...          0_none        7   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...          0_none       30   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "\n",
       "                                                    is_doublet_propagate  \\\n",
       "barcode_sample                                                             \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                     0   \n",
       "...                                                                  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "\n",
       "                                                    origin_M_F  \n",
       "barcode_sample                                                  \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                           F  \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                           M  \n",
       "...                                                        ...  \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...           M  \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "\n",
       "[74244 rows x 25 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading IDs of final cells to keep (after also excluding doublets by GEX, done in notebook M3)\n",
    "nodoublet_cell_IDs = pd.read_csv(save_path + 'obs_table_nodoublets.csv', index_col=0)\n",
    "nodoublet_cell_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74244 × 29058\n",
       "    obs: 'n_genes', 'donor', 'tissue_block', 'age', 'sample', 'percent_mito', 'n_counts', 'dataset', 'technique', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch', 'is_doublet'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subsetting the adata object to final cells\n",
    "adata = adata[list(nodoublet_cell_IDs.index), :].copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_genes\n",
      "donor\n",
      "tissue_block\n",
      "age\n",
      "sample\n",
      "percent_mito\n",
      "n_counts\n",
      "dataset\n",
      "technique\n",
      "scrublet_score\n",
      "scrublet_cluster_score\n",
      "bh_pval\n",
      "batch\n",
      "is_doublet\n",
      "barcode_sample_copy\n",
      "barcode\n",
      "souporcell_assignment\n",
      "inter_ind_doublet\n",
      "S_score\n",
      "G2M_score\n",
      "phase\n",
      "annotation_prev\n",
      "louvain\n",
      "is_doublet_propagate\n",
      "origin_M_F\n"
     ]
    }
   ],
   "source": [
    "# adding metadata\n",
    "for col in nodoublet_cell_IDs.columns:\n",
    "    print(col)\n",
    "    adata.obs[col] = nodoublet_cell_IDs.loc[adata.obs_names,col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to only invading trophoblast compartment\n",
    "# read in obs table with louvain labels from notebook S1 and subset to only clusters 6 and 8 which are VCT_CCC and all the invading troph\n",
    "obs_table_with_louvain = pd.read_csv(save_path + 'obs_table_adata_scvi_SN_and_multiome_with_louvain.csv', index_col=0)\n",
    "adata.obs['louvain_scvi'] = obs_table_with_louvain.loc[adata.obs_names, 'louvain_scvi_n_latent_17']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[adata.obs['louvain_scvi'].isin([6,8])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# figuring out how many PCs best explain variance in this data - to choose n_latent optimally\n",
    "# done in S1a notebook in jhub\n",
    "\n",
    "# optimal n_PCs = 18 here as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do some basic preprocessing\n",
    "adata.layers[\"raw_counts\"] = adata.X.copy() # preserve counts\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "adata.raw = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'donor' as categorical\n",
      "... storing 'tissue_block' as categorical\n",
      "... storing 'age' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'technique' as categorical\n",
      "... storing 'souporcell_assignment' as categorical\n",
      "... storing 'phase' as categorical\n",
      "... storing 'annotation_prev' as categorical\n",
      "... storing 'origin_M_F' as categorical\n",
      "/opt/conda/envs/scvi-singularity/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    }
   ],
   "source": [
    "# here we subset to HVGs, accounting for donor batch\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata,\n",
    "    n_top_genes=2000,\n",
    "    subset=True,\n",
    "    #layer=\"raw_counts\",\n",
    "    flavor=\"seurat\",\n",
    "    batch_key=\"sample\",\n",
    "    #batch_key=\"donor\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here you can specify your covariates you want to correct for (categorical or continuous)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m No batch_key inputted, assuming all cells are same batch                            \n",
      "\u001b[34mINFO    \u001b[0m No label_key inputted, assuming all cells have same label                           \n",
      "\u001b[34mINFO    \u001b[0m Using data from adata.layers\u001b[1m[\u001b[0m\u001b[32m\"raw_counts\"\u001b[0m\u001b[1m]\u001b[0m                                          \n",
      "\u001b[34mINFO    \u001b[0m Computing library size prior per batch                                              \n",
      "\u001b[34mINFO    \u001b[0m Successfully registered anndata object containing \u001b[1;36m5992\u001b[0m cells, \u001b[1;36m2000\u001b[0m vars, \u001b[1;36m1\u001b[0m batches, \n",
      "         \u001b[1;36m1\u001b[0m labels, and \u001b[1;36m0\u001b[0m proteins. Also registered \u001b[1;36m1\u001b[0m extra categorical covariates and \u001b[1;36m0\u001b[0m extra\n",
      "         continuous covariates.                                                              \n",
      "\u001b[34mINFO    \u001b[0m Please do not further modify adata until model is trained.                          \n"
     ]
    }
   ],
   "source": [
    "scvi.data.setup_anndata(\n",
    "    adata,\n",
    "    layer=\"raw_counts\",\n",
    "    categorical_covariate_keys=[\"sample\"],\n",
    "    #continuous_covariate_keys=[\"\"] # could try and regress n_genes in the future? not for now\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 09.11.2021 Running scVI after:\n",
    "\n",
    "- removing cell cycle-associated genes (calculated in M2 notebook)\n",
    "- exclusion of all doublets (scrublet and souporcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_latent_value 5\n",
      "n_latent_value 10\n",
      "n_latent_value 15\n",
      "n_latent_value 18\n",
      "n_latent_value 20\n"
     ]
    }
   ],
   "source": [
    "models = {}\n",
    "\n",
    "# just a few values here, it was 18 PCs I used in the conventional analysis\n",
    "n_latent_values = [5, 10, 15, \n",
    "                   18,\n",
    "                   20, \n",
    "                   #30, 40\n",
    "                  ]\n",
    "\n",
    "for n_latent_value in n_latent_values:\n",
    "    print('n_latent_value', n_latent_value)\n",
    "    models[n_latent_value] = scvi.model.SCVI(adata, n_latent = n_latent_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCVI Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">128</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">18</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.1</span>, dispersion: gene, \n",
       "gene_likelihood: zinb, latent_distribution: normal\n",
       "Training status: Not Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCVI Model with the following params: \n",
       "n_hidden: \u001b[1;36m128\u001b[0m, n_latent: \u001b[1;36m18\u001b[0m, n_layers: \u001b[1;36m1\u001b[0m, dropout_rate: \u001b[1;36m0.1\u001b[0m, dispersion: gene, \n",
       "gene_likelihood: zinb, latent_distribution: normal\n",
       "Training status: Not Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "\n",
       "To print summary of associated AnnData, use: <span style=\"color: #af00d7; text-decoration-color: #af00d7\">scvi.data.view_anndata_setup(model.adata)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n",
       "\n",
       "To print summary of associated AnnData, use: \u001b[38;5;128mscvi.data.view_anndata_setup(model.adata)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models[18]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 5\n",
      "Epoch 400/400: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 400/400 [02:43<00:00,  2.44it/s, loss=998, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 10\n",
      "Epoch 400/400: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 400/400 [02:43<00:00,  2.44it/s, loss=986, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 15\n",
      "Epoch 400/400: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 400/400 [02:43<00:00,  2.44it/s, loss=990, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 18\n",
      "Epoch 400/400: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 400/400 [02:43<00:00,  2.44it/s, loss=980, v_num=1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 20\n",
      "Epoch 400/400: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 400/400 [02:43<00:00,  2.44it/s, loss=967, v_num=1]\n"
     ]
    }
   ],
   "source": [
    "latent_representations = {}\n",
    "\n",
    "for n_latent_value in n_latent_values:\n",
    "    print('training model for n_latent_value:', n_latent_value)\n",
    "    models[n_latent_value].train()\n",
    "    \n",
    "    # get latent representation as a dataframe\n",
    "    latent_representations[n_latent_value] = models[n_latent_value].get_latent_representation()\n",
    "    \n",
    "    # add it to the adata object\n",
    "    adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)] = latent_representations[n_latent_value]\n",
    "    curr_df = pd.DataFrame(adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)])\n",
    "    \n",
    "    # save the latent representation\n",
    "    curr_df.to_csv('./20211109_obsm_with_scVI_latent_representation_n_' + str(n_latent_value) + '_after_CCG_removal_and_all_doublet_exclusion_by_sample_donor_P13_all_SN_and_multiome_only_invading_trophoblast.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rerunning for invading troph after excluding SCT contaminants (notebook S1a1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path = '/lustre/scratch117/cellgen/team292/aa22/adata_objects/202007_snRNA-seq_MFI/snRNA_seq_MFI_202007_adatas/202012_FINAL_reanalysis_with_souporcell_deconvolution_common_variants/donor_P13_all_samples_analysis_202110/'\n",
    "\n",
    "# Loading object with raw counts, filtered cells and genes object\n",
    "# before any doublet exclusion\n",
    "adata = sc.read(save_path + 'adata_raw_filtered.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>donor</th>\n",
       "      <th>tissue_block</th>\n",
       "      <th>age</th>\n",
       "      <th>sample</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>dataset</th>\n",
       "      <th>technique</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>...</th>\n",
       "      <th>barcode</th>\n",
       "      <th>souporcell_assignment</th>\n",
       "      <th>inter_ind_doublet</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>annotation_prev</th>\n",
       "      <th>louvain</th>\n",
       "      <th>is_doublet_propagate</th>\n",
       "      <th>origin_M_F</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGATACATG-1_WSSS_PLA8764121</th>\n",
       "      <td>1253</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1820.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.191617</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCAAGATACATG-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.053412</td>\n",
       "      <td>0.056947</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0_none</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAATCCCA-1_WSSS_PLA8764121</th>\n",
       "      <td>1279</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>1868.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.032895</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAAATCCCA-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.120729</td>\n",
       "      <td>-0.113529</td>\n",
       "      <td>G1</td>\n",
       "      <td>dM2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACTGTGT-1_WSSS_PLA8764121</th>\n",
       "      <td>473</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>556.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.091892</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAACTGTGT-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.042173</td>\n",
       "      <td>-0.027934</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAAGCTGCC-1_WSSS_PLA8764121</th>\n",
       "      <td>1360</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>2116.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.012335</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACAAGCTGCC-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.121624</td>\n",
       "      <td>0.050396</td>\n",
       "      <td>G2M</td>\n",
       "      <td>dNK1</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACATAACCCA-1_WSSS_PLA8764121</th>\n",
       "      <td>1891</td>\n",
       "      <td>P13</td>\n",
       "      <td>PU8-9_B5</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>WSSS_PLA8764121</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>3274.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.073634</td>\n",
       "      <td>...</td>\n",
       "      <td>AAACCCACATAACCCA-1_WSSS_PLA8764121</td>\n",
       "      <td>WSSS_PLA8764121_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.125064</td>\n",
       "      <td>-0.066023</td>\n",
       "      <td>G1</td>\n",
       "      <td>dS1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>2409</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.007270</td>\n",
       "      <td>5227.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033632</td>\n",
       "      <td>-0.079893</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>4129</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>17379.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.102190</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.059688</td>\n",
       "      <td>-0.098701</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>3116</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.004499</td>\n",
       "      <td>7779.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.036876</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.066281</td>\n",
       "      <td>-0.038885</td>\n",
       "      <td>S</td>\n",
       "      <td>0_none</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>3696</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>20078.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.031627</td>\n",
       "      <td>-0.068024</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_Pla_Camb10687916</th>\n",
       "      <td>2130</td>\n",
       "      <td>P13</td>\n",
       "      <td>unknown</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916</td>\n",
       "      <td>0.012930</td>\n",
       "      <td>5259.0</td>\n",
       "      <td>snRNA-seq</td>\n",
       "      <td>10X</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_...</td>\n",
       "      <td>Pla_Camb10714920_and_40110_Pla_Camb10687916_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.110820</td>\n",
       "      <td>-0.058170</td>\n",
       "      <td>G1</td>\n",
       "      <td>0_none</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74244 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    n_genes donor  \\\n",
       "barcode_sample                                                      \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                     1253   P13   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                     1279   P13   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                      473   P13   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                     1360   P13   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                     1891   P13   \n",
       "...                                                     ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...     2409   P13   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...     4129   P13   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...     3116   P13   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...     3696   P13   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...     2130   P13   \n",
       "\n",
       "                                                   tissue_block      age  \\\n",
       "barcode_sample                                                             \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                     PU8-9_B5  8-9_PCW   \n",
       "...                                                         ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...      unknown  8-9_PCW   \n",
       "\n",
       "                                                                                         sample  \\\n",
       "barcode_sample                                                                                    \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121   \n",
       "...                                                                                         ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916   \n",
       "\n",
       "                                                    percent_mito  n_counts  \\\n",
       "barcode_sample                                                               \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                      0.000000    1820.0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                      0.000535    1868.0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                      0.000000     556.0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                      0.000473    2116.0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                      0.000305    3274.0   \n",
       "...                                                          ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...      0.007270    5227.0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...      0.002532   17379.0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...      0.004499    7779.0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...      0.001992   20078.0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...      0.012930    5259.0   \n",
       "\n",
       "                                                      dataset technique  \\\n",
       "barcode_sample                                                            \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                  snRNA-seq       10X   \n",
       "...                                                       ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  snRNA-seq       10X   \n",
       "\n",
       "                                                    scrublet_score  ...  \\\n",
       "barcode_sample                                                      ...   \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                        0.191617  ...   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                        0.032895  ...   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                        0.091892  ...   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                        0.012335  ...   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                        0.073634  ...   \n",
       "...                                                            ...  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...        0.013699  ...   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...        0.102190  ...   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...        0.036876  ...   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...        0.109137  ...   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...        0.109137  ...   \n",
       "\n",
       "                                                                                              barcode  \\\n",
       "barcode_sample                                                                                          \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                 AAACCCAAGATACATG-1_WSSS_PLA8764121   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                 AAACCCACAAATCCCA-1_WSSS_PLA8764121   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                 AAACCCACAACTGTGT-1_WSSS_PLA8764121   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                 AAACCCACAAGCTGCC-1_WSSS_PLA8764121   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                 AAACCCACATAACCCA-1_WSSS_PLA8764121   \n",
       "...                                                                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_...   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_...   \n",
       "\n",
       "                                                                            souporcell_assignment  \\\n",
       "barcode_sample                                                                                      \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                              WSSS_PLA8764121_0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                              WSSS_PLA8764121_1   \n",
       "...                                                                                           ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...  Pla_Camb10714920_and_40110_Pla_Camb10687916_1   \n",
       "\n",
       "                                                    inter_ind_doublet  \\\n",
       "barcode_sample                                                          \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                  0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                  0   \n",
       "...                                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...                  0   \n",
       "\n",
       "                                                     S_score G2M_score phase  \\\n",
       "barcode_sample                                                                 \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                  0.053412  0.056947   G2M   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                 -0.120729 -0.113529    G1   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                  0.042173 -0.027934     S   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                 -0.121624  0.050396   G2M   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                 -0.125064 -0.066023    G1   \n",
       "...                                                      ...       ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...  0.033632 -0.079893     S   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P... -0.059688 -0.098701    G1   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...  0.066281 -0.038885     S   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P... -0.031627 -0.068024    G1   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P... -0.110820 -0.058170    G1   \n",
       "\n",
       "                                                   annotation_prev  louvain  \\\n",
       "barcode_sample                                                                \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                          0_none       16   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                             dM2       10   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                          0_none       14   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                            dNK1       14   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                             dS1        8   \n",
       "...                                                            ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...          0_none        7   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...          0_none       30   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...          0_none        2   \n",
       "\n",
       "                                                    is_doublet_propagate  \\\n",
       "barcode_sample                                                             \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                                     0   \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                                     0   \n",
       "...                                                                  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...                     0   \n",
       "\n",
       "                                                    origin_M_F  \n",
       "barcode_sample                                                  \n",
       "AAACCCAAGATACATG-1_WSSS_PLA8764121                           F  \n",
       "AAACCCACAAATCCCA-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACAACTGTGT-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACAAGCTGCC-1_WSSS_PLA8764121                           M  \n",
       "AAACCCACATAACCCA-1_WSSS_PLA8764121                           M  \n",
       "...                                                        ...  \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920_and_40110_P...           M  \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920_and_40110_P...           F  \n",
       "\n",
       "[74244 rows x 25 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading IDs of final cells to keep (after also excluding doublets by GEX, done in notebook M3)\n",
    "nodoublet_cell_IDs = pd.read_csv(save_path + 'obs_table_nodoublets.csv', index_col=0)\n",
    "nodoublet_cell_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74244 × 29058\n",
       "    obs: 'n_genes', 'donor', 'tissue_block', 'age', 'sample', 'percent_mito', 'n_counts', 'dataset', 'technique', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch', 'is_doublet'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subsetting the adata object to final cells\n",
    "adata = adata[list(nodoublet_cell_IDs.index), :].copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_genes\n",
      "donor\n",
      "tissue_block\n",
      "age\n",
      "sample\n",
      "percent_mito\n",
      "n_counts\n",
      "dataset\n",
      "technique\n",
      "scrublet_score\n",
      "scrublet_cluster_score\n",
      "bh_pval\n",
      "batch\n",
      "is_doublet\n",
      "barcode_sample_copy\n",
      "barcode\n",
      "souporcell_assignment\n",
      "inter_ind_doublet\n",
      "S_score\n",
      "G2M_score\n",
      "phase\n",
      "annotation_prev\n",
      "louvain\n",
      "is_doublet_propagate\n",
      "origin_M_F\n"
     ]
    }
   ],
   "source": [
    "# adding metadata\n",
    "for col in nodoublet_cell_IDs.columns:\n",
    "    print(col)\n",
    "    adata.obs[col] = nodoublet_cell_IDs.loc[adata.obs_names,col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to only invading trophoblast compartment\n",
    "# read in obs table with louvain labels from notebook S1 and subset to only clusters 6 and 8 which are VCT_CCC and all the invading troph\n",
    "obs_table_with_louvain = pd.read_csv(save_path + 'obs_table_adata_scvi_SN_and_multiome_with_louvain_invading_troph_only.csv', index_col=0)\n",
    "# subset to only inv troph\n",
    "adata = adata[obs_table_with_louvain.index,:].copy()\n",
    "\n",
    "adata.obs['louvain_scvi'] = obs_table_with_louvain.loc[adata.obs_names, 'louvain_scvi_n_latent_18']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clusters 2, 6 and 7 are SCT contaminants\n",
    "np.unique(adata.obs['louvain_scvi'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[-adata.obs['louvain_scvi'].isin([2,6,7])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 3, 4, 5, 8, 9])"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(adata.obs['louvain_scvi'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# figuring out how many PCs best explain variance in this data - to choose n_latent optimally\n",
    "# done in S1a notebook in jhub\n",
    "\n",
    "# optimal n_PCs = 18 here as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do some basic preprocessing\n",
    "adata.layers[\"raw_counts\"] = adata.X.copy() # preserve counts\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "adata.raw = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'donor' as categorical\n",
      "... storing 'tissue_block' as categorical\n",
      "... storing 'age' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'technique' as categorical\n",
      "... storing 'souporcell_assignment' as categorical\n",
      "... storing 'phase' as categorical\n",
      "... storing 'annotation_prev' as categorical\n",
      "... storing 'origin_M_F' as categorical\n",
      "/opt/conda/envs/scvi-singularity/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    }
   ],
   "source": [
    "# here we subset to HVGs, accounting for donor batch\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata,\n",
    "    n_top_genes=2000,\n",
    "    subset=True,\n",
    "    #layer=\"raw_counts\",\n",
    "    flavor=\"seurat\",\n",
    "    batch_key=\"sample\",\n",
    "    #batch_key=\"donor\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here you can specify your covariates you want to correct for (categorical or continuous)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m No batch_key inputted, assuming all cells are same batch                            \n",
      "\u001b[34mINFO    \u001b[0m No label_key inputted, assuming all cells have same label                           \n",
      "\u001b[34mINFO    \u001b[0m Using data from adata.layers\u001b[1m[\u001b[0m\u001b[32m\"raw_counts\"\u001b[0m\u001b[1m]\u001b[0m                                          \n",
      "\u001b[34mINFO    \u001b[0m Computing library size prior per batch                                              \n",
      "\u001b[34mINFO    \u001b[0m Successfully registered anndata object containing \u001b[1;36m4626\u001b[0m cells, \u001b[1;36m2000\u001b[0m vars, \u001b[1;36m1\u001b[0m batches, \n",
      "         \u001b[1;36m1\u001b[0m labels, and \u001b[1;36m0\u001b[0m proteins. Also registered \u001b[1;36m1\u001b[0m extra categorical covariates and \u001b[1;36m0\u001b[0m extra\n",
      "         continuous covariates.                                                              \n",
      "\u001b[34mINFO    \u001b[0m Please do not further modify adata until model is trained.                          \n"
     ]
    }
   ],
   "source": [
    "scvi.data.setup_anndata(\n",
    "    adata,\n",
    "    layer=\"raw_counts\",\n",
    "    categorical_covariate_keys=[\"sample\"],\n",
    "    #continuous_covariate_keys=[\"\"] # could try and regress n_genes in the future? not for now\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 09.11.2021 Running scVI after:\n",
    "\n",
    "- removing cell cycle-associated genes (calculated in M2 notebook)\n",
    "- exclusion of all doublets (scrublet and souporcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_latent_value 18\n"
     ]
    }
   ],
   "source": [
    "models = {}\n",
    "\n",
    "# just a few values here, it was 18 PCs I used in the conventional analysis\n",
    "n_latent_values = [18]\n",
    "\n",
    "for n_latent_value in n_latent_values:\n",
    "    print('n_latent_value', n_latent_value)\n",
    "    models[n_latent_value] = scvi.model.SCVI(adata, n_latent = n_latent_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCVI Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">128</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">18</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.1</span>, dispersion: gene, \n",
       "gene_likelihood: zinb, latent_distribution: normal\n",
       "Training status: Not Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCVI Model with the following params: \n",
       "n_hidden: \u001b[1;36m128\u001b[0m, n_latent: \u001b[1;36m18\u001b[0m, n_layers: \u001b[1;36m1\u001b[0m, dropout_rate: \u001b[1;36m0.1\u001b[0m, dispersion: gene, \n",
       "gene_likelihood: zinb, latent_distribution: normal\n",
       "Training status: Not Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "\n",
       "To print summary of associated AnnData, use: <span style=\"color: #af00d7; text-decoration-color: #af00d7\">scvi.data.view_anndata_setup(model.adata)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n",
       "\n",
       "To print summary of associated AnnData, use: \u001b[38;5;128mscvi.data.view_anndata_setup(model.adata)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models[18]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True, used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training model for n_latent_value: 18\n",
      "Epoch 400/400: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 400/400 [02:07<00:00,  3.14it/s, loss=1.02e+03, v_num=1]\n"
     ]
    }
   ],
   "source": [
    "latent_representations = {}\n",
    "\n",
    "for n_latent_value in n_latent_values:\n",
    "    print('training model for n_latent_value:', n_latent_value)\n",
    "    models[n_latent_value].train()\n",
    "    \n",
    "    # get latent representation as a dataframe\n",
    "    latent_representations[n_latent_value] = models[n_latent_value].get_latent_representation()\n",
    "    \n",
    "    # add it to the adata object\n",
    "    adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)] = latent_representations[n_latent_value]\n",
    "    curr_df = pd.DataFrame(adata.obsm[\"X_scVI_n_latent_\" + str(n_latent_value)])\n",
    "    \n",
    "    # save the latent representation\n",
    "    curr_df.to_csv('./20211109_obsm_with_scVI_latent_representation_n_' + str(n_latent_value) + '_after_CCG_removal_and_all_doublet_exclusion_by_sample_donor_P13_all_SN_and_multiome_only_invading_trophoblast_cleanup_1.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
