{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DE analysis for CellPhnoDB\n",
    "\n",
    "### With `limma`\n",
    "\n",
    "MFI atlas: all donors invading trophoblast\n",
    "\n",
    "Here using cells/nuclei assignment as a categorical covariate as it is the biggest source of gene coverage variation\n",
    "\n",
    "**The DE gene lists obtained here are further used in CellPhoneDB analysis and TF analysis**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "from glob import iglob\n",
    "import anndata\n",
    "import os\n",
    "#import sklearn\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "import matplotlib as mpl\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absl-py==0.12.0\n",
      "aiohttp==3.7.4.post0\n",
      "airr==1.3.1\n",
      "alabaster==0.7.12\n",
      "alembic @ file:///home/conda/feedstock_root/build_artifacts/alembic_1613901514078/work\n",
      "anndata @ file:///home/conda/feedstock_root/build_artifacts/anndata_1605539061264/work\n",
      "annoy @ file:///home/conda/feedstock_root/build_artifacts/python-annoy_1610271511811/work\n",
      "anyio @ file:///home/conda/feedstock_root/build_artifacts/anyio_1614388751160/work/dist\n",
      "arboreto==0.1.6\n",
      "argon2-cffi @ file:///home/conda/feedstock_root/build_artifacts/argon2-cffi_1610522574055/work\n",
      "async-generator==1.10\n",
      "async-timeout==3.0.1\n",
      "attrs @ file:///home/conda/feedstock_root/build_artifacts/attrs_1605083924122/work\n",
      "Babel @ file:///home/conda/feedstock_root/build_artifacts/babel_1605182336601/work\n",
      "backcall @ file:///home/conda/feedstock_root/build_artifacts/backcall_1592338393461/work\n",
      "backports.functools-lru-cache==1.6.1\n",
      "bbknn @ file:///opt/conda/conda-bld/bbknn_1616434096000/work\n",
      "bleach @ file:///home/conda/feedstock_root/build_artifacts/bleach_1612213472466/work\n",
      "blinker==1.4\n",
      "bokeh==2.3.0\n",
      "boltons==20.2.1\n",
      "brotlipy==0.7.0\n",
      "cached-property @ file:///home/conda/feedstock_root/build_artifacts/cached_property_1615209429212/work\n",
      "cachetools==4.2.1\n",
      "certifi==2021.10.8\n",
      "certipy==0.1.3\n",
      "cffi @ file:///home/conda/feedstock_root/build_artifacts/cffi_1613413861439/work\n",
      "chardet @ file:///home/conda/feedstock_root/build_artifacts/chardet_1610093490430/work\n",
      "click==7.1.2\n",
      "cloudpickle==1.6.0\n",
      "colorama @ file:///home/conda/feedstock_root/build_artifacts/colorama_1602866480661/work\n",
      "commonmark==0.9.1\n",
      "conda==4.9.2\n",
      "conda-pack @ file:///home/conda/feedstock_root/build_artifacts/conda-pack_1642622318035/work\n",
      "conda-package-handling @ file:///home/conda/feedstock_root/build_artifacts/conda-package-handling_1602876795040/work\n",
      "cryptography @ file:///home/conda/feedstock_root/build_artifacts/cryptography_1612993729974/work\n",
      "cycler==0.10.0\n",
      "Cython @ file:///home/conda/feedstock_root/build_artifacts/cython_1613867105642/work\n",
      "cytoolz==0.11.0\n",
      "dask==2021.3.1\n",
      "decorator==4.4.2\n",
      "defusedxml==0.6.0\n",
      "dill==0.3.3\n",
      "distributed==2021.3.1\n",
      "docutils @ file:///home/conda/feedstock_root/build_artifacts/docutils_1610127755164/work\n",
      "entrypoints @ file:///home/conda/feedstock_root/build_artifacts/entrypoints_1605121927639/work/dist/entrypoints-0.3-py2.py3-none-any.whl\n",
      "et-xmlfile==1.0.1\n",
      "fbpca==1.0\n",
      "frozendict==1.2\n",
      "fsspec==0.8.7\n",
      "future==0.18.2\n",
      "geosketch==1.2\n",
      "get-version==2.1\n",
      "google-auth==1.28.0\n",
      "google-auth-oauthlib==0.4.3\n",
      "grpcio==1.36.1\n",
      "h5py @ file:///home/conda/feedstock_root/build_artifacts/h5py_1604753633781/work\n",
      "HeapDict==1.0.1\n",
      "idna @ file:///home/conda/feedstock_root/build_artifacts/idna_1593328102638/work\n",
      "imageio==2.9.0\n",
      "imagesize==1.2.0\n",
      "importlib-metadata @ file:///home/conda/feedstock_root/build_artifacts/importlib-metadata_1614193722981/work\n",
      "interlap==0.2.7\n",
      "intervaltree==2.1.0\n",
      "ipykernel @ file:///home/conda/feedstock_root/build_artifacts/ipykernel_1613852938555/work/dist/ipykernel-5.5.0-py3-none-any.whl\n",
      "ipython @ file:///home/conda/feedstock_root/build_artifacts/ipython_1614393573755/work\n",
      "ipython-genutils==0.2.0\n",
      "ipywidgets @ file:///home/conda/feedstock_root/build_artifacts/ipywidgets_1609995587151/work\n",
      "jedi @ file:///home/conda/feedstock_root/build_artifacts/jedi_1610146791023/work\n",
      "Jinja2 @ file:///home/conda/feedstock_root/build_artifacts/jinja2_1612119311452/work\n",
      "joblib @ file:///home/conda/feedstock_root/build_artifacts/joblib_1612898609989/work\n",
      "json5 @ file:///home/conda/feedstock_root/build_artifacts/json5_1600692310011/work\n",
      "jsonschema @ file:///home/conda/feedstock_root/build_artifacts/jsonschema_1602551949684/work\n",
      "jupyter-client @ file:///home/conda/feedstock_root/build_artifacts/jupyter_client_1610375432619/work\n",
      "jupyter-core @ file:///home/conda/feedstock_root/build_artifacts/jupyter_core_1612125275706/work\n",
      "jupyter-packaging @ file:///home/conda/feedstock_root/build_artifacts/jupyter-packaging_1613054948399/work\n",
      "jupyter-resource-usage @ file:///home/conda/feedstock_root/build_artifacts/jupyter-resource-usage_1608550684149/work\n",
      "jupyter-rsession-proxy==1.2\n",
      "jupyter-server @ file:///home/conda/feedstock_root/build_artifacts/jupyter_server_1614028749313/work\n",
      "jupyter-server-proxy==3.0.2\n",
      "jupyter-telemetry @ file:///home/conda/feedstock_root/build_artifacts/jupyter_telemetry_1605173804246/work\n",
      "jupyterhub @ file:///home/conda/feedstock_root/build_artifacts/jupyterhub-feedstock_1614255305026/work\n",
      "jupyterlab @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_1614041538214/work\n",
      "jupyterlab-pygments @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_pygments_1601375948261/work\n",
      "jupyterlab-server @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_server_1613760084674/work\n",
      "jupyterlab-widgets @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_widgets_1609173350931/work\n",
      "kiwisolver @ file:///home/conda/feedstock_root/build_artifacts/kiwisolver_1610099769230/work\n",
      "legacy-api-wrap==0.0.0\n",
      "leidenalg @ file:///home/conda/feedstock_root/build_artifacts/leidenalg_1605889147051/work\n",
      "llvmlite==0.34.0\n",
      "locket==0.2.1\n",
      "loompy==3.0.6\n",
      "louvain @ file:///home/conda/feedstock_root/build_artifacts/louvain_1609939902418/work\n",
      "Mako @ file:///home/conda/feedstock_root/build_artifacts/mako_1610659158978/work\n",
      "mamba @ file:///home/conda/feedstock_root/build_artifacts/mamba_1613160772800/work\n",
      "Markdown==3.3.4\n",
      "MarkupSafe @ file:///home/conda/feedstock_root/build_artifacts/markupsafe_1610127565888/work\n",
      "matplotlib @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-suite_1611858708474/work\n",
      "mistune @ file:///home/conda/feedstock_root/build_artifacts/mistune_1610112875388/work\n",
      "mock @ file:///home/conda/feedstock_root/build_artifacts/mock_1610094566888/work\n",
      "msgpack==1.0.2\n",
      "multidict==5.1.0\n",
      "multiprocessing-on-dill==3.5.0a4\n",
      "natsort @ file:///home/conda/feedstock_root/build_artifacts/natsort_1611580267114/work\n",
      "nbclassic @ file:///home/conda/feedstock_root/build_artifacts/nbclassic_1610352513187/work\n",
      "nbclient @ file:///home/conda/feedstock_root/build_artifacts/nbclient_1614336084111/work\n",
      "nbconvert @ file:///home/conda/feedstock_root/build_artifacts/nbconvert_1605401836768/work\n",
      "nbformat @ file:///home/conda/feedstock_root/build_artifacts/nbformat_1611005694671/work\n",
      "nbresuse==0.3.6\n",
      "nest-asyncio @ file:///home/conda/feedstock_root/build_artifacts/nest-asyncio_1605195931949/work\n",
      "networkx @ file:///home/conda/feedstock_root/build_artifacts/networkx_1598210780226/work\n",
      "notebook @ file:///home/conda/feedstock_root/build_artifacts/notebook_1610575313697/work\n",
      "numba @ file:///home/conda/feedstock_root/build_artifacts/numba_1599084802945/work\n",
      "numexpr @ file:///home/conda/feedstock_root/build_artifacts/numexpr_1614971340455/work\n",
      "numpy @ file:///home/conda/feedstock_root/build_artifacts/numpy_1612953005674/work\n",
      "numpy-groupies==0.9.13\n",
      "oauthlib==3.0.1\n",
      "olefile @ file:///home/conda/feedstock_root/build_artifacts/olefile_1602866521163/work\n",
      "openpyxl==3.0.7\n",
      "opt-einsum==3.3.0\n",
      "packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1612459636436/work\n",
      "pamela==1.0.0\n",
      "pandas==1.2.3\n",
      "pandocfilters==1.4.2\n",
      "parasail==1.2.4\n",
      "parso @ file:///home/conda/feedstock_root/build_artifacts/parso_1607618318316/work\n",
      "partd==1.1.0\n",
      "patsy==0.5.1\n",
      "pexpect @ file:///home/conda/feedstock_root/build_artifacts/pexpect_1602535608087/work\n",
      "pickleshare @ file:///home/conda/feedstock_root/build_artifacts/pickleshare_1602536217715/work\n",
      "Pillow @ file:///home/conda/feedstock_root/build_artifacts/pillow_1615243974661/work\n",
      "plotly==4.14.3\n",
      "prometheus-client @ file:///home/conda/feedstock_root/build_artifacts/prometheus_client_1605543085815/work\n",
      "prompt-toolkit @ file:///home/conda/feedstock_root/build_artifacts/prompt-toolkit_1613068548426/work\n",
      "protobuf==3.15.6\n",
      "psutil @ file:///home/conda/feedstock_root/build_artifacts/psutil_1610127095720/work\n",
      "ptyprocess @ file:///home/conda/feedstock_root/build_artifacts/ptyprocess_1609419310487/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl\n",
      "pyarrow==0.16.0\n",
      "pyasn1==0.4.8\n",
      "pyasn1-modules==0.2.8\n",
      "pycairo==1.20.0\n",
      "pycosat @ file:///home/conda/feedstock_root/build_artifacts/pycosat_1610094800877/work\n",
      "pycparser @ file:///home/conda/feedstock_root/build_artifacts/pycparser_1593275161868/work\n",
      "pycurl==7.43.0.6\n",
      "Pygments @ file:///home/conda/feedstock_root/build_artifacts/pygments_1613412098605/work\n",
      "PyJWT @ file:///home/conda/feedstock_root/build_artifacts/pyjwt_1610910308735/work\n",
      "pynndescent @ file:///home/conda/feedstock_root/build_artifacts/pynndescent_1612844832048/work\n",
      "pyOpenSSL @ file:///home/conda/feedstock_root/build_artifacts/pyopenssl_1608055815057/work\n",
      "pyparsing==2.4.7\n",
      "PyQt5==5.12.3\n",
      "PyQt5-sip==4.19.18\n",
      "PyQtChart==5.12\n",
      "PyQtWebEngine==5.12.1\n",
      "pyro-api==0.1.2\n",
      "pyro-ppl==1.6.0\n",
      "pyrsistent @ file:///home/conda/feedstock_root/build_artifacts/pyrsistent_1610146798212/work\n",
      "pysam==0.16.0.1\n",
      "pyscenic==0.11.0\n",
      "PySocks @ file:///home/conda/feedstock_root/build_artifacts/pysocks_1610291447907/work\n",
      "python-dateutil==2.8.1\n",
      "python-editor==1.0.4\n",
      "python-igraph @ file:///home/conda/feedstock_root/build_artifacts/python-igraph_1609945946184/work\n",
      "python-json-logger @ file:///home/conda/feedstock_root/build_artifacts/python-json-logger_1602545356084/work\n",
      "python-Levenshtein==0.12.2\n",
      "pytoml==0.1.21\n",
      "pytorch-lightning==1.2.5\n",
      "pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1612179539967/work\n",
      "PyWavelets==1.1.1\n",
      "PyYAML==5.3.1\n",
      "pyzmq @ file:///home/conda/feedstock_root/build_artifacts/pyzmq_1613478739884/work\n",
      "requests @ file:///home/conda/feedstock_root/build_artifacts/requests_1608156231189/work\n",
      "requests-oauthlib==1.3.0\n",
      "retrying==1.3.3\n",
      "rich==10.0.0\n",
      "rpy2==3.4.3\n",
      "rsa==4.7.2\n",
      "ruamel-yaml-conda @ file:///home/conda/feedstock_root/build_artifacts/ruamel_yaml_1611943339799/work\n",
      "ruamel.yaml @ file:///home/conda/feedstock_root/build_artifacts/ruamel.yaml_1610291375472/work\n",
      "ruamel.yaml.clib @ file:///home/conda/feedstock_root/build_artifacts/ruamel.yaml.clib_1610146840614/work\n",
      "scanorama==1.7.1\n",
      "scanpy @ file:///opt/conda/conda-bld/scanpy_1614155786839/work\n",
      "scikit-image==0.18.1\n",
      "scikit-learn @ file:///home/conda/feedstock_root/build_artifacts/scikit-learn_1611079738450/work\n",
      "scipy @ file:///home/conda/feedstock_root/build_artifacts/scipy_1615955928702/work\n",
      "scirpy==0.6.1\n",
      "scrublet==0.2.3\n",
      "scvelo==0.2.3\n",
      "scvi-tools==0.9.1\n",
      "seaborn @ file:///home/conda/feedstock_root/build_artifacts/seaborn-split_1611834504644/work\n",
      "Send2Trash==1.5.0\n",
      "setuptools-scm @ file:///home/conda/feedstock_root/build_artifacts/setuptools_scm_1615996969755/work\n",
      "simpervisor==0.4\n",
      "sinfo @ file:///home/conda/feedstock_root/build_artifacts/sinfo_1597764960136/work\n",
      "six @ file:///home/conda/feedstock_root/build_artifacts/six_1590081179328/work\n",
      "sniffio @ file:///home/conda/feedstock_root/build_artifacts/sniffio_1610318319305/work\n",
      "snowballstemmer @ file:///home/conda/feedstock_root/build_artifacts/snowballstemmer_1611270869511/work\n",
      "sortedcontainers==2.3.0\n",
      "Sphinx @ file:///home/conda/feedstock_root/build_artifacts/sphinx_1616256380770/work\n",
      "sphinxcontrib-applehelp==1.0.2\n",
      "sphinxcontrib-devhelp==1.0.2\n",
      "sphinxcontrib-htmlhelp==1.0.3\n",
      "sphinxcontrib-jsmath==1.0.1\n",
      "sphinxcontrib-qthelp==1.0.3\n",
      "sphinxcontrib-serializinghtml==1.1.4\n",
      "SQLAlchemy @ file:///home/conda/feedstock_root/build_artifacts/sqlalchemy_1612225077951/work\n",
      "squarify==0.4.3\n",
      "statsmodels @ file:///home/conda/feedstock_root/build_artifacts/statsmodels_1612273599609/work\n",
      "stdlib-list @ file:///home/conda/feedstock_root/build_artifacts/stdlib-list_1602130019264/work\n",
      "tables @ file:///home/conda/feedstock_root/build_artifacts/pytables_1603317058579/work\n",
      "tblib==1.7.0\n",
      "tensorboard==2.4.1\n",
      "tensorboard-plugin-wit==1.8.0\n",
      "terminado @ file:///home/conda/feedstock_root/build_artifacts/terminado_1609794172745/work\n",
      "testpath==0.4.4\n",
      "texttable @ file:///home/conda/feedstock_root/build_artifacts/texttable_1599404337529/work\n",
      "threadpoolctl @ file:///tmp/tmp79xdzxkt/threadpoolctl-2.1.0-py3-none-any.whl\n",
      "tifffile==2021.3.17\n",
      "toml @ file:///home/conda/feedstock_root/build_artifacts/toml_1604308577558/work\n",
      "toolz==0.11.1\n",
      "torch==1.8.1\n",
      "torchmetrics==0.2.0\n",
      "tornado @ file:///home/conda/feedstock_root/build_artifacts/tornado_1610094706440/work\n",
      "tqdm @ file:///home/conda/feedstock_root/build_artifacts/tqdm_1614279180094/work\n",
      "traitlets @ file:///home/conda/feedstock_root/build_artifacts/traitlets_1602771532708/work\n",
      "typing-extensions==3.7.4.3\n",
      "tzlocal==2.1\n",
      "umap-learn @ file:///home/conda/feedstock_root/build_artifacts/umap-learn_1642213339708/work\n",
      "urllib3 @ file:///home/conda/feedstock_root/build_artifacts/urllib3_1611695416663/work\n",
      "velocyto==0.17.17\n",
      "wcwidth @ file:///home/conda/feedstock_root/build_artifacts/wcwidth_1600965781394/work\n",
      "webencodings==0.5.1\n",
      "Werkzeug==1.0.1\n",
      "widgetsnbextension @ file:///home/conda/feedstock_root/build_artifacts/widgetsnbextension_1605475534911/work\n",
      "yamlordereddictloader==0.4.0\n",
      "yarl==1.6.3\n",
      "zict==2.0.0\n",
      "zipp @ file:///home/conda/feedstock_root/build_artifacts/zipp_1603668650351/work\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pip freeze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: If you miss a compact list, please try `print_header`!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.7.5\n",
      "scanpy      1.7.1\n",
      "sinfo       0.3.1\n",
      "-----\n",
      "PIL                 8.1.2\n",
      "anndata             0.7.5\n",
      "anyio               NA\n",
      "attr                20.3.0\n",
      "babel               2.9.0\n",
      "backcall            0.2.0\n",
      "brotli              NA\n",
      "cairo               1.20.0\n",
      "certifi             2021.10.08\n",
      "cffi                1.14.5\n",
      "chardet             4.0.0\n",
      "cloudpickle         1.6.0\n",
      "colorama            0.4.4\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "cytoolz             0.11.0\n",
      "dask                2021.03.1\n",
      "dateutil            2.8.1\n",
      "decorator           4.4.2\n",
      "fsspec              0.8.7\n",
      "get_version         2.1\n",
      "google              NA\n",
      "h5py                3.1.0\n",
      "idna                2.10\n",
      "igraph              0.8.3\n",
      "ipykernel           5.5.0\n",
      "ipython_genutils    0.2.0\n",
      "ipywidgets          7.6.3\n",
      "jedi                0.18.0\n",
      "jinja2              2.11.3\n",
      "joblib              1.0.1\n",
      "json5               NA\n",
      "jsonschema          3.2.0\n",
      "jupyter_server      1.4.1\n",
      "jupyterlab_server   2.3.0\n",
      "kiwisolver          1.3.1\n",
      "legacy_api_wrap     0.0.0\n",
      "leidenalg           0.8.3\n",
      "llvmlite            0.34.0\n",
      "louvain             0.7.0\n",
      "markupsafe          1.1.1\n",
      "matplotlib          3.3.4\n",
      "mpl_toolkits        NA\n",
      "natsort             7.1.1\n",
      "nbclassic           NA\n",
      "nbformat            5.1.2\n",
      "numba               0.51.2\n",
      "numexpr             2.7.3\n",
      "numpy               1.20.1\n",
      "packaging           20.9\n",
      "pandas              1.2.3\n",
      "parso               0.8.1\n",
      "pexpect             4.8.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "prometheus_client   NA\n",
      "prompt_toolkit      3.0.16\n",
      "psutil              5.8.0\n",
      "ptyprocess          0.7.0\n",
      "pvectorc            NA\n",
      "pyarrow             0.16.0\n",
      "pygments            2.8.0\n",
      "pyparsing           2.4.7\n",
      "pyrsistent          NA\n",
      "pytoml              NA\n",
      "pytz                2021.1\n",
      "requests            2.25.1\n",
      "ruamel              NA\n",
      "scanpy              1.7.1\n",
      "scipy               1.6.1\n",
      "seaborn             0.11.1\n",
      "send2trash          NA\n",
      "setuptools_scm      NA\n",
      "sinfo               0.3.1\n",
      "six                 1.15.0\n",
      "sklearn             0.24.1\n",
      "sniffio             1.2.0\n",
      "socks               1.7.1\n",
      "sphinxcontrib       NA\n",
      "statsmodels         0.12.2\n",
      "storemagic          NA\n",
      "tables              3.6.1\n",
      "tblib               1.7.0\n",
      "texttable           1.6.3\n",
      "tlz                 0.11.0\n",
      "toolz               0.11.1\n",
      "tornado             6.1\n",
      "traitlets           5.0.5\n",
      "typing_extensions   NA\n",
      "urllib3             1.26.3\n",
      "wcwidth             0.2.5\n",
      "yaml                5.3.1\n",
      "zmq                 22.0.3\n",
      "-----\n",
      "IPython             7.21.0\n",
      "jupyter_client      6.1.11\n",
      "jupyter_core        4.7.1\n",
      "jupyterlab          3.0.9\n",
      "notebook            6.2.0\n",
      "-----\n",
      "Python 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27) [GCC 9.3.0]\n",
      "Linux-4.15.0-158-generic-x86_64-with-glibc2.10\n",
      "60 logical CPU cores, x86_64\n",
      "-----\n",
      "Session information updated at 2022-03-01 14:59\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80)  # low dpi (dots per inch) yields small inline figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path = '/lustre/scratch117/cellgen/team292/aa22/adata_objects/202111_MFI_sc_sn_new_scVI_analysis/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading raw filtered data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(save_path + 'adata_raw_filtered.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dS2               65746\n",
       "dS1               42586\n",
       "SCT               28177\n",
       "VCT               23472\n",
       "uSMC              17043\n",
       "dEpi_secretory    16834\n",
       "dM1               14099\n",
       "dS3               13197\n",
       "dNK2              12487\n",
       "HOFB              10733\n",
       "dNK1              10450\n",
       "fF1               10420\n",
       "PV                 9980\n",
       "dT_cells           9740\n",
       "dM2                9221\n",
       "Endo_M             6877\n",
       "VCT_p              6679\n",
       "EVT_1              5289\n",
       "NK                 5187\n",
       "iEVT               3676\n",
       "VCT_CCC            3373\n",
       "MO                 3050\n",
       "dNK3               2952\n",
       "ILC3               2511\n",
       "EVT_2              2358\n",
       "T_cells            2126\n",
       "VCT_fusing         1971\n",
       "dT_regs            1862\n",
       "Endo_L             1761\n",
       "Endo_F             1466\n",
       "fF2                1396\n",
       "M3                 1299\n",
       "B_cells             774\n",
       "DC                  697\n",
       "dDC                 694\n",
       "Plasma              255\n",
       "Granulocytes        195\n",
       "dEpi_lumenal        135\n",
       "eEVT                 28\n",
       "GC                   19\n",
       "Name: cell_type, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in final annotations, subset to only inv troph + VCT + VCT_p and all the final 'cell_type' labels\n",
    "final_annot = pd.read_csv(save_path + 'final_annotation_all_cells_and_nulcei_20211123.csv', index_col=0)\n",
    "final_annot['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "adata = adata[final_annot.index,:]\n",
    "adata.obs['cell_type'] = final_annot.loc[adata.obs_names,'cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to only cell states of interest to test DEGs\n",
    "adata = adata[adata.obs['cell_type'].isin(['VCT','VCT_p','VCT_CCC','EVT_1','EVT_2','eEVT','iEVT','GC'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 44894 × 30800\n",
       "    obs: 'n_genes', 'sample', 'technology', 'tissue', 'dev_age', 'donor', 'dataset', 'run', 'number_of_individuals_multiplexed', 'batch', 'percent_mito', 'n_counts', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'is_doublet', 'cell_type'\n",
       "    var: 'gene_ids-0', 'feature_types-0', 'genome-0', 'n_cells-0', 'gene_ids-1', 'feature_types-1', 'genome-1', 'n_cells-1', 'gene_ids-10', 'feature_types-10', 'genome-10', 'n_cells-10', 'gene_ids-11', 'feature_types-11', 'genome-11', 'n_cells-11', 'gene_ids-12', 'feature_types-12', 'genome-12', 'n_cells-12', 'gene_ids-13', 'feature_types-13', 'genome-13', 'n_cells-13', 'gene_ids-14', 'feature_types-14', 'genome-14', 'n_cells-14', 'gene_ids-15', 'feature_types-15', 'genome-15', 'n_cells-15', 'gene_ids-16', 'feature_types-16', 'genome-16', 'n_cells-16', 'gene_ids-17', 'feature_types-17', 'genome-17', 'n_cells-17', 'gene_ids-18', 'feature_types-18', 'genome-18', 'n_cells-18', 'gene_ids-19', 'feature_types-19', 'genome-19', 'n_cells-19', 'gene_ids-2', 'feature_types-2', 'genome-2', 'n_cells-2', 'gene_ids-20', 'feature_types-20', 'genome-20', 'n_cells-20', 'gene_ids-21', 'feature_types-21', 'genome-21', 'n_cells-21', 'gene_ids-22', 'feature_types-22', 'genome-22', 'n_cells-22', 'gene_ids-23', 'feature_types-23', 'genome-23', 'n_cells-23', 'gene_ids-24', 'feature_types-24', 'genome-24', 'n_cells-24', 'gene_ids-25', 'feature_types-25', 'genome-25', 'n_cells-25', 'gene_ids-26', 'feature_types-26', 'genome-26', 'n_cells-26', 'gene_ids-27', 'feature_types-27', 'genome-27', 'n_cells-27', 'gene_ids-28', 'feature_types-28', 'genome-28', 'n_cells-28', 'gene_ids-29', 'feature_types-29', 'genome-29', 'n_cells-29', 'gene_ids-3', 'feature_types-3', 'genome-3', 'n_cells-3', 'gene_ids-30', 'feature_types-30', 'genome-30', 'n_cells-30', 'gene_ids-31', 'feature_types-31', 'genome-31', 'n_cells-31', 'gene_ids-32', 'feature_types-32', 'genome-32', 'n_cells-32', 'gene_ids-33', 'feature_types-33', 'genome-33', 'n_cells-33', 'gene_ids-34', 'feature_types-34', 'genome-34', 'n_cells-34', 'gene_ids-35', 'feature_types-35', 'genome-35', 'n_cells-35', 'gene_ids-36', 'feature_types-36', 'genome-36', 'n_cells-36', 'gene_ids-37', 'feature_types-37', 'genome-37', 'n_cells-37', 'gene_ids-38', 'feature_types-38', 'genome-38', 'n_cells-38', 'gene_ids-39', 'feature_types-39', 'genome-39', 'n_cells-39', 'gene_ids-4', 'feature_types-4', 'genome-4', 'n_cells-4', 'gene_ids-40', 'feature_types-40', 'genome-40', 'n_cells-40', 'gene_ids-41', 'feature_types-41', 'genome-41', 'n_cells-41', 'gene_ids-42', 'feature_types-42', 'genome-42', 'n_cells-42', 'gene_ids-43', 'feature_types-43', 'genome-43', 'n_cells-43', 'gene_ids-44', 'feature_types-44', 'genome-44', 'n_cells-44', 'gene_ids-45', 'feature_types-45', 'genome-45', 'n_cells-45', 'gene_ids-46', 'feature_types-46', 'genome-46', 'n_cells-46', 'gene_ids-47', 'feature_types-47', 'genome-47', 'n_cells-47', 'gene_ids-48', 'feature_types-48', 'genome-48', 'n_cells-48', 'gene_ids-49', 'feature_types-49', 'genome-49', 'n_cells-49', 'gene_ids-5', 'feature_types-5', 'genome-5', 'n_cells-5', 'gene_ids-50', 'feature_types-50', 'genome-50', 'n_cells-50', 'gene_ids-51', 'feature_types-51', 'genome-51', 'n_cells-51', 'gene_ids-52', 'feature_types-52', 'genome-52', 'n_cells-52', 'gene_ids-53', 'feature_types-53', 'genome-53', 'n_cells-53', 'gene_ids-54', 'feature_types-54', 'genome-54', 'n_cells-54', 'gene_ids-55', 'feature_types-55', 'genome-55', 'n_cells-55', 'gene_ids-56', 'feature_types-56', 'genome-56', 'n_cells-56', 'gene_ids-57', 'feature_types-57', 'genome-57', 'n_cells-57', 'gene_ids-58', 'feature_types-58', 'genome-58', 'n_cells-58', 'gene_ids-59', 'feature_types-59', 'genome-59', 'n_cells-59', 'gene_ids-6', 'feature_types-6', 'genome-6', 'n_cells-6', 'gene_ids-60', 'feature_types-60', 'genome-60', 'n_cells-60', 'gene_ids-61', 'feature_types-61', 'genome-61', 'n_cells-61', 'gene_ids-7', 'feature_types-7', 'genome-7', 'n_cells-7', 'gene_ids-8', 'feature_types-8', 'genome-8', 'n_cells-8', 'gene_ids-9', 'feature_types-9', 'genome-9', 'n_cells-9'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____________________________________________________________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DE with `limma`\n",
    "\n",
    "Separately for the invading trophoblast in a retrograde manner for the trajectory:\n",
    "- VCT_CCC vs [VCT_p + VCT]\n",
    "- EVT_1 vs VCT_CCC\n",
    "- EVT_2 vs EVT_1\n",
    "- eEVT vs EVT_2\n",
    "- iEVT vs EVT_2\n",
    "- GC vs iEVT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "VCT        23472\n",
       "VCT_p       6679\n",
       "EVT_1       5289\n",
       "iEVT        3676\n",
       "VCT_CCC     3373\n",
       "EVT_2       2358\n",
       "eEVT          28\n",
       "GC            19\n",
       "Name: cell_type, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "normalizing by total count per cell\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "    finished (0:00:04): normalized adata.X and added    'n_counts', counts per cell before normalization (adata.obs)\n"
     ]
    }
   ],
   "source": [
    "# normalise and log transform values for limma\n",
    "sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# saving normalised and log transformed values into the raw attribute to later be able to plot\n",
    "adata.raw = adata.copy()\n",
    "\n",
    "adata.var_names_make_unique()\n",
    "adata.obs_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(adata.obs['cell_type']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for the purpose of first comparison here uniting VCT and VCT_p\n",
    "adata.obs['cell_type'] = ['VCT' if ct in ['VCT','VCT_p'] else ct for ct in adata.obs['cell_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['EVT_1', 'EVT_2', 'GC', 'VCT', 'VCT_CCC', 'eEVT', 'iEVT'],\n",
       "       dtype=object),\n",
       " array([ 5289,  2358,    19, 30151,  3373,    28,  3676]))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(adata.obs['cell_type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "# downsample 'VCT' about tenfold - otherwise limma will break\n",
    "VCT_indices = list(adata[adata.obs['cell_type'] == 'VCT'].obs_names)\n",
    "adata_downsampled_VCT = adata[VCT_indices[:3000],:].copy()\n",
    "adata_no_VCT = adata[adata.obs['cell_type'] != 'VCT'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14743"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(adata_no_VCT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata_no_VCT.concatenate(adata_downsampled_VCT, join='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>sample</th>\n",
       "      <th>technology</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dev_age</th>\n",
       "      <th>donor</th>\n",
       "      <th>dataset</th>\n",
       "      <th>run</th>\n",
       "      <th>number_of_individuals_multiplexed</th>\n",
       "      <th>batch</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>scrublet_cluster_score</th>\n",
       "      <th>bh_pval</th>\n",
       "      <th>is_doublet</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CTGATAGAGCCCAGCT-1_FCA7167219-0</th>\n",
       "      <td>4070</td>\n",
       "      <td>FCA7167219</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>13_PCW</td>\n",
       "      <td>6</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.027002</td>\n",
       "      <td>14851.0</td>\n",
       "      <td>0.186441</td>\n",
       "      <td>0.118367</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>False</td>\n",
       "      <td>EVT_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTACTCTCTGCGGCA-1_FCA7167221-0</th>\n",
       "      <td>655</td>\n",
       "      <td>FCA7167221</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>11_PCW</td>\n",
       "      <td>7</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003621</td>\n",
       "      <td>1657.0</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>0.167267</td>\n",
       "      <td>0.242677</td>\n",
       "      <td>False</td>\n",
       "      <td>iEVT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGATAGAGCCCAGCT-1_FCA7167221-0</th>\n",
       "      <td>412</td>\n",
       "      <td>FCA7167221</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>11_PCW</td>\n",
       "      <td>7</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.044892</td>\n",
       "      <td>646.0</td>\n",
       "      <td>0.090426</td>\n",
       "      <td>0.167267</td>\n",
       "      <td>0.242677</td>\n",
       "      <td>False</td>\n",
       "      <td>EVT_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGAATCCACTCTGTC-1_FCA7167222-0</th>\n",
       "      <td>1320</td>\n",
       "      <td>FCA7167222</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_non_immune</td>\n",
       "      <td>11_PCW</td>\n",
       "      <td>7</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>3820.0</td>\n",
       "      <td>0.051118</td>\n",
       "      <td>0.055810</td>\n",
       "      <td>0.874083</td>\n",
       "      <td>False</td>\n",
       "      <td>EVT_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTTTCTAGATAGGAG-1_FCA7167222-0</th>\n",
       "      <td>1076</td>\n",
       "      <td>FCA7167222</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_non_immune</td>\n",
       "      <td>11_PCW</td>\n",
       "      <td>7</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.012456</td>\n",
       "      <td>3372.0</td>\n",
       "      <td>0.060870</td>\n",
       "      <td>0.055810</td>\n",
       "      <td>0.874083</td>\n",
       "      <td>False</td>\n",
       "      <td>iEVT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACATCAGCAGCGTAAG-1_FCA7474064-1</th>\n",
       "      <td>4031</td>\n",
       "      <td>FCA7474064</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>placenta</td>\n",
       "      <td>7_PCW</td>\n",
       "      <td>10</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>25892.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.033059</td>\n",
       "      <td>20932.0</td>\n",
       "      <td>0.094972</td>\n",
       "      <td>0.131034</td>\n",
       "      <td>0.824150</td>\n",
       "      <td>False</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACATCAGCAGGGAGAG-1_FCA7474064-1</th>\n",
       "      <td>2755</td>\n",
       "      <td>FCA7474064</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>placenta</td>\n",
       "      <td>7_PCW</td>\n",
       "      <td>10</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>25892.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.053909</td>\n",
       "      <td>9516.0</td>\n",
       "      <td>0.098712</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.874976</td>\n",
       "      <td>False</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACATCAGCATCGGAAG-1_FCA7474064-1</th>\n",
       "      <td>4900</td>\n",
       "      <td>FCA7474064</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>placenta</td>\n",
       "      <td>7_PCW</td>\n",
       "      <td>10</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>25892.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.047498</td>\n",
       "      <td>29938.0</td>\n",
       "      <td>0.209756</td>\n",
       "      <td>0.139812</td>\n",
       "      <td>0.824150</td>\n",
       "      <td>False</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACATCAGGTCAGCTAT-1_FCA7474064-1</th>\n",
       "      <td>3507</td>\n",
       "      <td>FCA7474064</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>placenta</td>\n",
       "      <td>7_PCW</td>\n",
       "      <td>10</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>25892.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.082652</td>\n",
       "      <td>12474.0</td>\n",
       "      <td>0.047161</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.824150</td>\n",
       "      <td>False</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACATCAGGTCCGAATT-1_FCA7474064-1</th>\n",
       "      <td>3809</td>\n",
       "      <td>FCA7474064</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>placenta</td>\n",
       "      <td>7_PCW</td>\n",
       "      <td>10</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>25892.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.093463</td>\n",
       "      <td>17729.0</td>\n",
       "      <td>0.075795</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>0.824150</td>\n",
       "      <td>False</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17743 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 n_genes      sample        technology  \\\n",
       "barcode_sample                                                           \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0     4070  FCA7167219  10X_3'_scRNA-seq   \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0      655  FCA7167221  10X_3'_scRNA-seq   \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0      412  FCA7167221  10X_3'_scRNA-seq   \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0     1320  FCA7167222  10X_3'_scRNA-seq   \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0     1076  FCA7167222  10X_3'_scRNA-seq   \n",
       "...                                  ...         ...               ...   \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1     4031  FCA7474064  10X_3'_scRNA-seq   \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1     2755  FCA7474064  10X_3'_scRNA-seq   \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1     4900  FCA7474064  10X_3'_scRNA-seq   \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1     3507  FCA7474064  10X_3'_scRNA-seq   \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1     3809  FCA7474064  10X_3'_scRNA-seq   \n",
       "\n",
       "                                             tissue dev_age donor  \\\n",
       "barcode_sample                                                      \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0      decidua_immune  13_PCW     6   \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0      decidua_immune  11_PCW     7   \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0      decidua_immune  11_PCW     7   \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0  decidua_non_immune  11_PCW     7   \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0  decidua_non_immune  11_PCW     7   \n",
       "...                                             ...     ...   ...   \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1            placenta   7_PCW    10   \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1            placenta   7_PCW    10   \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1            placenta   7_PCW    10   \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1            placenta   7_PCW    10   \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1            placenta   7_PCW    10   \n",
       "\n",
       "                                      dataset      run  \\\n",
       "barcode_sample                                           \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0  Vento_Nature      NaN   \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0  Vento_Nature      NaN   \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0  Vento_Nature      NaN   \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0  Vento_Nature      NaN   \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0  Vento_Nature      NaN   \n",
       "...                                       ...      ...   \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1  Vento_Nature  25892.0   \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1  Vento_Nature  25892.0   \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1  Vento_Nature  25892.0   \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1  Vento_Nature  25892.0   \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1  Vento_Nature  25892.0   \n",
       "\n",
       "                                 number_of_individuals_multiplexed batch  \\\n",
       "barcode_sample                                                             \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0                                  1     0   \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0                                  1     0   \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0                                  1     0   \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0                                  1     0   \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0                                  1     0   \n",
       "...                                                            ...   ...   \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1                                  1     1   \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1                                  1     1   \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1                                  1     1   \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1                                  1     1   \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1                                  1     1   \n",
       "\n",
       "                                 percent_mito  n_counts  scrublet_score  \\\n",
       "barcode_sample                                                            \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0      0.027002   14851.0        0.186441   \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0      0.003621    1657.0        0.175000   \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0      0.044892     646.0        0.090426   \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0      0.000524    3820.0        0.051118   \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0      0.012456    3372.0        0.060870   \n",
       "...                                       ...       ...             ...   \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1      0.033059   20932.0        0.094972   \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1      0.053909    9516.0        0.098712   \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1      0.047498   29938.0        0.209756   \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1      0.082652   12474.0        0.047161   \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1      0.093463   17729.0        0.075795   \n",
       "\n",
       "                                 scrublet_cluster_score   bh_pval  is_doublet  \\\n",
       "barcode_sample                                                                  \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0                0.118367  0.857143       False   \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0                0.167267  0.242677       False   \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0                0.167267  0.242677       False   \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0                0.055810  0.874083       False   \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0                0.055810  0.874083       False   \n",
       "...                                                 ...       ...         ...   \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1                0.131034  0.824150       False   \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1                0.111111  0.874976       False   \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1                0.139812  0.824150       False   \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1                0.142857  0.824150       False   \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1                0.156250  0.824150       False   \n",
       "\n",
       "                                cell_type  \n",
       "barcode_sample                             \n",
       "CTGATAGAGCCCAGCT-1_FCA7167219-0     EVT_2  \n",
       "ATTACTCTCTGCGGCA-1_FCA7167221-0      iEVT  \n",
       "CTGATAGAGCCCAGCT-1_FCA7167221-0     EVT_2  \n",
       "CAGAATCCACTCTGTC-1_FCA7167222-0     EVT_1  \n",
       "CCTTTCTAGATAGGAG-1_FCA7167222-0      iEVT  \n",
       "...                                   ...  \n",
       "ACATCAGCAGCGTAAG-1_FCA7474064-1       VCT  \n",
       "ACATCAGCAGGGAGAG-1_FCA7474064-1       VCT  \n",
       "ACATCAGCATCGGAAG-1_FCA7474064-1       VCT  \n",
       "ACATCAGGTCAGCTAT-1_FCA7474064-1       VCT  \n",
       "ACATCAGGTCCGAATT-1_FCA7474064-1       VCT  \n",
       "\n",
       "[17743 rows x 17 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['EVT_1', 'EVT_2', 'GC', 'VCT', 'VCT_CCC', 'eEVT', 'iEVT'],\n",
       "       dtype=object),\n",
       " array([5289, 2358,   19, 3000, 3373,   28, 3676]))"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(adata.obs['cell_type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"10X_3'_scRNA-seq\", \"10X_3'_snRNA-seq\", '10X_multiome'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(adata.obs['technology'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# going to use origin (SN or SC) as covariate here\n",
    "adata.obs['limma_cov_sn_or_sc'] = ['sn' if technology in [\"10X_3'_snRNA-seq\", '10X_multiome'] else 'sc' for technology in adata.obs['technology']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EVT_1\n",
      "sc    3947\n",
      "sn    1342\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n",
      "EVT_2\n",
      "sn    1210\n",
      "sc    1148\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n",
      "GC\n",
      "sn    19\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n",
      "VCT\n",
      "sc    3000\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n",
      "VCT_CCC\n",
      "sc    2268\n",
      "sn    1105\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n",
      "eEVT\n",
      "sn    26\n",
      "sc     2\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n",
      "iEVT\n",
      "sn    2597\n",
      "sc    1079\n",
      "Name: limma_cov_sn_or_sc, dtype: int64\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for ct in np.unique(adata.obs['cell_type']):\n",
    "    print(ct)\n",
    "    curr_subset = adata[adata.obs['cell_type'] == ct]\n",
    "    print(curr_subset.obs['limma_cov_sn_or_sc'].value_counts())\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset adatas by comparison\n",
    "# and repeating this for others..\n",
    "\n",
    "adata_subset = adata[adata.obs['cell_type'].isin(['VCT','VCT_CCC'])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "# marker calling\n",
    "\n",
    "#t = adata_hvg.X.toarray().T\n",
    "t = adata_subset.X.toarray().T\n",
    "df = pd.DataFrame(data=t, columns= adata_subset.obs.index, index=adata_subset.var_names)\n",
    "\n",
    "meta_df = pd.DataFrame(data={'Cell':list(adata_subset.obs.index),\n",
    "                             'cell_type':[ str(i) for i in adata_subset.obs['cell_type']],\n",
    "                             'sn_or_sc':[ str(i) for i in adata_subset.obs['limma_cov_sn_or_sc']],\n",
    "                             #'sample':[ str(i) for i in adata_subset.obs['sample']]\n",
    "                            })\n",
    "meta_df.set_index('Cell', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "outpath = './limma_DEGs/20211123/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(limma)\n",
    "library(edgeR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 4.0.4 (2021-02-15)\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "Running under: Ubuntu 20.04.1 LTS\n",
      "\n",
      "Matrix products: default\n",
      "BLAS/LAPACK: /opt/conda/lib/libopenblasp-r0.3.12.so\n",
      "\n",
      "locale:\n",
      " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
      " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
      " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
      " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
      " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
      "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
      "\n",
      "attached base packages:\n",
      "[1] tools     stats     graphics  grDevices utils     datasets  methods  \n",
      "[8] base     \n",
      "\n",
      "other attached packages:\n",
      "[1] edgeR_3.32.1 limma_3.46.0\n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_4.0.4  Rcpp_1.0.7      grid_4.0.4      locfit_1.5-9.4 \n",
      "[5] lattice_0.20-41\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6373"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(df.columns) & set(meta_df['Cell']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['VCT', 'VCT_CCC'], dtype=object), array([3000, 3373]))"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(meta_df['cell_type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "case = 'VCT_CCC'\n",
    "ctrl = 'VCT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"VCT_CCC\" \"VCT\"    \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 6373 shared cells\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"case\" \"ctrl\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i df -i meta_df -i outpath -i ctrl -i case \n",
    "\n",
    "library(limma)\n",
    "library(edgeR)\n",
    "\n",
    "# Format\n",
    "ex_mat=as.matrix(df)\n",
    "rownames(meta_df) = meta_df$Cell\n",
    "\n",
    "# subset meta\n",
    "meta_df = subset(meta_df, cell_type %in% unlist(c(ctrl, case)) )\n",
    "print(unique(meta_df$cell_type))\n",
    "\n",
    "# Shared cells\n",
    "shared_cells = intersect(rownames(meta_df), colnames(ex_mat))\n",
    "message(length(shared_cells), ' shared cells')\n",
    "ex_mat = ex_mat[, shared_cells]\n",
    "meta_df = meta_df[shared_cells,]\n",
    "\n",
    "#print(dim(ex_mat))\n",
    "\n",
    "# Filter lowly expressed genes\n",
    "keep = rowSums(ex_mat, na.rm=T) > 0.1\n",
    "ex_mat = ex_mat[ keep, ]\n",
    "keep = aveLogCPM(ex_mat) > 0.1\n",
    "ex_mat = ex_mat[ keep, ]\n",
    "\n",
    "#print(dim(ex_mat))\n",
    "\n",
    "# Extract celltypes\n",
    "cells = rownames(meta_df)\n",
    "celltypes = unique(meta_df$cell_type)\n",
    "covariates = meta_df$sn_or_sc\n",
    "\n",
    "# Extract cells in cluster and rest\n",
    "cells_case = rownames(subset(meta_df, cell_type == case))\n",
    "cells_ctrl = rownames(subset(meta_df, cell_type == ctrl)) # changed from control to ctrl\n",
    "\n",
    "# build cluster_type vector\n",
    "cluster_type = rep(0, length(cells))\n",
    "names(cluster_type) = cells\n",
    "cluster_type[ cells_case ] = 'case'\n",
    "cluster_type[ cells_ctrl ] = 'ctrl'\n",
    "\n",
    "print(unique(cluster_type))\n",
    "\n",
    "design.matrix <- model.matrix(~ 0 + cluster_type + covariates)\n",
    "#design.matrix <- model.matrix(~ 0 + cluster_type)\n",
    "\n",
    "# Now tell limma how do you want to compare (i.e. case vs control)\n",
    "contrast.matrix <- makeContrasts(caseVScontrol = cluster_typecase - cluster_typectrl, levels = design.matrix)\n",
    "\n",
    "# Make model and run contrasts\n",
    "fit <- lmFit(ex_mat, design.matrix)\n",
    "fit <- contrasts.fit(fit, contrast.matrix)\n",
    "fit <- eBayes(fit)\n",
    "\n",
    "# Make a dataframe containing the important data\n",
    "results = topTable(fit, adjust=\"fdr\", number = nrow(ex_mat), coef = 'caseVScontrol')\n",
    "\n",
    "# Add and filter needed data\n",
    "results$Gene = rownames(results)\n",
    "results = results[ , c('Gene', 'logFC', 'P.Value', 'adj.P.Val')]\n",
    "results$AveExpr_cluster = apply(ex_mat[ results$Gene, cells_case], 1, mean)\n",
    "results$AveExpr_rest = apply(ex_mat[ results$Gene, cells_ctrl], 1, mean)\n",
    "results$percentExpr_cluster = apply(ex_mat[ results$Gene, cells_case], 1, function(x) sum(c(x > 0)+0) ) / length(cells_case)\n",
    "results$percentExpr_rest = apply(ex_mat[ results$Gene, cells_ctrl], 1, function(x) sum(c(x > 0)+0) ) / length(cells_ctrl)\n",
    "\n",
    "results$AveExpr_cluster = round(results$AveExpr_cluster, 6)\n",
    "results$AveExpr_rest = round(results$AveExpr_rest, 6)\n",
    "results$percentExpr_cluster = round(results$percentExpr_cluster, 6)\n",
    "results$percentExpr_rest = round(results$percentExpr_rest, 6)\n",
    "# and store it as csv file\n",
    "write.csv(results, file = paste0(outpath, \n",
    "                                 '20211123_', case, '_vs_', ctrl, '_limma_DEGs.csv'), row.names = F, col.names = T, quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in limma tables, filter genes by adj.P.Val (taking all with < 0.05) and format according to what CellPhone needs for further use in cell-cell comm analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCT_CCC_vs_VCT\n",
      "4058 significant upreg DEGs here \n",
      "\n",
      "EVT_1_vs_VCT_CCC\n",
      "4603 significant upreg DEGs here \n",
      "\n",
      "EVT_2_vs_EVT_1\n",
      "2513 significant upreg DEGs here \n",
      "\n",
      "eEVT_vs_EVT_2\n",
      "2148 significant upreg DEGs here \n",
      "\n",
      "iEVT_vs_EVT_2\n",
      "4876 significant upreg DEGs here \n",
      "\n",
      "GC_vs_iEVT\n",
      "560 significant upreg DEGs here \n",
      "\n"
     ]
    }
   ],
   "source": [
    "DEG_tables = {}\n",
    "\n",
    "comparisons = ['VCT_CCC_vs_VCT',\n",
    "              'EVT_1_vs_VCT_CCC',\n",
    "              'EVT_2_vs_EVT_1',\n",
    "              'eEVT_vs_EVT_2',\n",
    "              'iEVT_vs_EVT_2',\n",
    "              'GC_vs_iEVT']\n",
    "\n",
    "\n",
    "for comparison in comparisons:\n",
    "    print(comparison)\n",
    "    \n",
    "    # main call type for which this DE list is\n",
    "    main_ct = comparison.split('_vs_')[0]\n",
    "    \n",
    "    DEG_tables[comparison] = pd.read_csv('./limma_DEGs/20211123/20211123_' + comparison + '_limma_DEGs.csv', #index_col=0\n",
    "                                        )\n",
    "    # filter by FDR\n",
    "    DEG_tables[comparison] = DEG_tables[comparison][DEG_tables[comparison]['adj.P.Val'] < 0.05]\n",
    "    # filter by logFC (>0)\n",
    "    DEG_tables[comparison] = DEG_tables[comparison][DEG_tables[comparison]['logFC'] > 0]\n",
    "    \n",
    "    # add 'cluster' labels\n",
    "    DEG_tables[comparison]['cluster'] = [main_ct for idx in DEG_tables[comparison].index]\n",
    "    \n",
    "    DEG_tables[comparison].columns = ['gene' if column == 'Gene' else column for column in DEG_tables[comparison].columns]\n",
    "    \n",
    "    DEG_tables[comparison] = DEG_tables[comparison].loc[:,['cluster','gene','logFC', 'P.Value', 'adj.P.Val', 'AveExpr_cluster',\n",
    "       'AveExpr_rest', 'percentExpr_cluster', 'percentExpr_rest']]\n",
    "\n",
    "    # additionally (02.12.2021): filtering genes that are expr in <10% of cells in cluster (to be consistent with cellphone)\n",
    "    #DEG_tables[comparison] = DEG_tables[comparison][DEG_tables[comparison]['percentExpr_cluster'] >= 0.1]\n",
    "    \n",
    "    \n",
    "    print(DEG_tables[comparison].shape[0],'significant upreg DEGs here \\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCT_CCC_vs_VCT\n",
      "2855 significant upreg DEGs here \n",
      "\n",
      "EVT_1_vs_VCT_CCC\n",
      "3116 significant upreg DEGs here \n",
      "\n",
      "EVT_2_vs_EVT_1\n",
      "1433 significant upreg DEGs here \n",
      "\n",
      "eEVT_vs_EVT_2\n",
      "1479 significant upreg DEGs here \n",
      "\n",
      "iEVT_vs_EVT_2\n",
      "3870 significant upreg DEGs here \n",
      "\n",
      "GC_vs_iEVT\n",
      "266 significant upreg DEGs here \n",
      "\n"
     ]
    }
   ],
   "source": [
    "DEG_tables = {}\n",
    "\n",
    "comparisons = ['VCT_CCC_vs_VCT',\n",
    "              'EVT_1_vs_VCT_CCC',\n",
    "              'EVT_2_vs_EVT_1',\n",
    "              'eEVT_vs_EVT_2',\n",
    "              'iEVT_vs_EVT_2',\n",
    "              'GC_vs_iEVT']\n",
    "\n",
    "\n",
    "for comparison in comparisons:\n",
    "    print(comparison)\n",
    "    \n",
    "    # main call type for which this DE list is\n",
    "    main_ct = comparison.split('_vs_')[0]\n",
    "    \n",
    "    DEG_tables[comparison] = pd.read_csv('./limma_DEGs/20211123/20211123_' + comparison + '_limma_DEGs.csv', #index_col=0\n",
    "                                        )\n",
    "    # filter by FDR\n",
    "    DEG_tables[comparison] = DEG_tables[comparison][DEG_tables[comparison]['adj.P.Val'] < 0.05]\n",
    "    # filter by logFC (>0)\n",
    "    DEG_tables[comparison] = DEG_tables[comparison][DEG_tables[comparison]['logFC'] > 0]\n",
    "    \n",
    "    # add 'cluster' labels\n",
    "    DEG_tables[comparison]['cluster'] = [main_ct for idx in DEG_tables[comparison].index]\n",
    "    \n",
    "    DEG_tables[comparison].columns = ['gene' if column == 'Gene' else column for column in DEG_tables[comparison].columns]\n",
    "    \n",
    "    DEG_tables[comparison] = DEG_tables[comparison].loc[:,['cluster','gene','logFC', 'P.Value', 'adj.P.Val', 'AveExpr_cluster',\n",
    "       'AveExpr_rest', 'percentExpr_cluster', 'percentExpr_rest']]\n",
    "\n",
    "    # additionally (02.12.2021): filtering genes that are expr in <10% of cells in cluster (to be consistent with cellphone)\n",
    "    DEG_tables[comparison] = DEG_tables[comparison][DEG_tables[comparison]['percentExpr_cluster'] >= 0.1]\n",
    "    \n",
    "    \n",
    "    print(DEG_tables[comparison].shape[0],'significant upreg DEGs here \\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['cluster', 'gene', 'logFC', 'P.Value', 'adj.P.Val', 'AveExpr_cluster',\n",
       "       'AveExpr_rest', 'percentExpr_cluster', 'percentExpr_rest'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DEG_tables['eEVT_vs_EVT_2'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>EIF4E1B</td>\n",
       "      <td>0.589620</td>\n",
       "      <td>2.542969e-300</td>\n",
       "      <td>6.267909e-296</td>\n",
       "      <td>0.593823</td>\n",
       "      <td>0.001864</td>\n",
       "      <td>0.607143</td>\n",
       "      <td>0.002969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>NCAM1</td>\n",
       "      <td>3.614782</td>\n",
       "      <td>9.408643e-295</td>\n",
       "      <td>1.159521e-290</td>\n",
       "      <td>3.795461</td>\n",
       "      <td>0.105632</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.066582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>SULT1A2</td>\n",
       "      <td>0.802209</td>\n",
       "      <td>4.286850e-250</td>\n",
       "      <td>3.522076e-246</td>\n",
       "      <td>0.810678</td>\n",
       "      <td>0.006605</td>\n",
       "      <td>0.642857</td>\n",
       "      <td>0.007634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>TNFRSF8</td>\n",
       "      <td>0.952954</td>\n",
       "      <td>1.627569e-239</td>\n",
       "      <td>1.002908e-235</td>\n",
       "      <td>0.968758</td>\n",
       "      <td>0.009481</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.010178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>ADORA1</td>\n",
       "      <td>1.202695</td>\n",
       "      <td>2.035514e-233</td>\n",
       "      <td>1.003427e-229</td>\n",
       "      <td>1.227871</td>\n",
       "      <td>0.015473</td>\n",
       "      <td>0.785714</td>\n",
       "      <td>0.015691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2231</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>ZNF552</td>\n",
       "      <td>0.160193</td>\n",
       "      <td>4.467550e-03</td>\n",
       "      <td>4.933521e-02</td>\n",
       "      <td>0.260883</td>\n",
       "      <td>0.072702</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.078032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2232</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>KLHDC10</td>\n",
       "      <td>0.281399</td>\n",
       "      <td>4.504171e-03</td>\n",
       "      <td>4.971733e-02</td>\n",
       "      <td>0.697828</td>\n",
       "      <td>0.261461</td>\n",
       "      <td>0.678571</td>\n",
       "      <td>0.230704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2233</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>MTR</td>\n",
       "      <td>0.257660</td>\n",
       "      <td>4.512153e-03</td>\n",
       "      <td>4.978315e-02</td>\n",
       "      <td>0.553450</td>\n",
       "      <td>0.204410</td>\n",
       "      <td>0.607143</td>\n",
       "      <td>0.196353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2234</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>SDCBP2</td>\n",
       "      <td>0.051753</td>\n",
       "      <td>4.515229e-03</td>\n",
       "      <td>4.978437e-02</td>\n",
       "      <td>0.060622</td>\n",
       "      <td>0.006639</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.007634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2235</th>\n",
       "      <td>eEVT</td>\n",
       "      <td>GRB7</td>\n",
       "      <td>0.229916</td>\n",
       "      <td>4.516304e-03</td>\n",
       "      <td>4.978437e-02</td>\n",
       "      <td>0.391364</td>\n",
       "      <td>0.170506</td>\n",
       "      <td>0.464286</td>\n",
       "      <td>0.201442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1479 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster     gene     logFC        P.Value      adj.P.Val  \\\n",
       "0       eEVT  EIF4E1B  0.589620  2.542969e-300  6.267909e-296   \n",
       "1       eEVT    NCAM1  3.614782  9.408643e-295  1.159521e-290   \n",
       "2       eEVT  SULT1A2  0.802209  4.286850e-250  3.522076e-246   \n",
       "3       eEVT  TNFRSF8  0.952954  1.627569e-239  1.002908e-235   \n",
       "4       eEVT   ADORA1  1.202695  2.035514e-233  1.003427e-229   \n",
       "...      ...      ...       ...            ...            ...   \n",
       "2231    eEVT   ZNF552  0.160193   4.467550e-03   4.933521e-02   \n",
       "2232    eEVT  KLHDC10  0.281399   4.504171e-03   4.971733e-02   \n",
       "2233    eEVT      MTR  0.257660   4.512153e-03   4.978315e-02   \n",
       "2234    eEVT   SDCBP2  0.051753   4.515229e-03   4.978437e-02   \n",
       "2235    eEVT     GRB7  0.229916   4.516304e-03   4.978437e-02   \n",
       "\n",
       "      AveExpr_cluster  AveExpr_rest  percentExpr_cluster  percentExpr_rest  \n",
       "0            0.593823      0.001864             0.607143          0.002969  \n",
       "1            3.795461      0.105632             1.000000          0.066582  \n",
       "2            0.810678      0.006605             0.642857          0.007634  \n",
       "3            0.968758      0.009481             0.750000          0.010178  \n",
       "4            1.227871      0.015473             0.785714          0.015691  \n",
       "...               ...           ...                  ...               ...  \n",
       "2231         0.260883      0.072702             0.357143          0.078032  \n",
       "2232         0.697828      0.261461             0.678571          0.230704  \n",
       "2233         0.553450      0.204410             0.607143          0.196353  \n",
       "2234         0.060622      0.006639             0.107143          0.007634  \n",
       "2235         0.391364      0.170506             0.464286          0.201442  \n",
       "\n",
       "[1479 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DEG_tables['eEVT_vs_EVT_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine all tables and save for CellPhone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# without any filtering\n",
    "joint_DE_table = pd.concat(DEG_tables.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>VCT_CCC</td>\n",
       "      <td>TPM1</td>\n",
       "      <td>1.456768</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.017393</td>\n",
       "      <td>0.758216</td>\n",
       "      <td>0.882597</td>\n",
       "      <td>0.594000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>VCT_CCC</td>\n",
       "      <td>FXYD5</td>\n",
       "      <td>0.655155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.518204</td>\n",
       "      <td>0.053708</td>\n",
       "      <td>0.510821</td>\n",
       "      <td>0.074667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>VCT_CCC</td>\n",
       "      <td>LDHA</td>\n",
       "      <td>1.148616</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.727698</td>\n",
       "      <td>1.231422</td>\n",
       "      <td>0.743848</td>\n",
       "      <td>0.732667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>VCT_CCC</td>\n",
       "      <td>FABP5</td>\n",
       "      <td>0.824986</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.645517</td>\n",
       "      <td>0.083226</td>\n",
       "      <td>0.454788</td>\n",
       "      <td>0.120000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>VCT_CCC</td>\n",
       "      <td>SLC16A3</td>\n",
       "      <td>0.864479</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.897528</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.640972</td>\n",
       "      <td>0.251667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>GC</td>\n",
       "      <td>PGR</td>\n",
       "      <td>0.135100</td>\n",
       "      <td>0.001135</td>\n",
       "      <td>0.048031</td>\n",
       "      <td>0.167571</td>\n",
       "      <td>0.026268</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.030468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>GC</td>\n",
       "      <td>AC244090.1</td>\n",
       "      <td>0.112713</td>\n",
       "      <td>0.001160</td>\n",
       "      <td>0.048682</td>\n",
       "      <td>0.128027</td>\n",
       "      <td>0.033475</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.067737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>GC</td>\n",
       "      <td>EPHA1-AS1</td>\n",
       "      <td>0.177610</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.048682</td>\n",
       "      <td>0.248932</td>\n",
       "      <td>0.052469</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.066921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>GC</td>\n",
       "      <td>MED31</td>\n",
       "      <td>0.178074</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.048682</td>\n",
       "      <td>0.228414</td>\n",
       "      <td>0.081405</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.143090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635</th>\n",
       "      <td>GC</td>\n",
       "      <td>KALRN</td>\n",
       "      <td>0.253556</td>\n",
       "      <td>0.001191</td>\n",
       "      <td>0.049684</td>\n",
       "      <td>0.403670</td>\n",
       "      <td>0.109891</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.129489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13019 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster        gene     logFC   P.Value  adj.P.Val  AveExpr_cluster  \\\n",
       "10   VCT_CCC        TPM1  1.456768  0.000000   0.000000         2.017393   \n",
       "14   VCT_CCC       FXYD5  0.655155  0.000000   0.000000         0.518204   \n",
       "21   VCT_CCC        LDHA  1.148616  0.000000   0.000000         1.727698   \n",
       "24   VCT_CCC       FABP5  0.824986  0.000000   0.000000         0.645517   \n",
       "26   VCT_CCC     SLC16A3  0.864479  0.000000   0.000000         0.897528   \n",
       "..       ...         ...       ...       ...        ...              ...   \n",
       "626       GC         PGR  0.135100  0.001135   0.048031         0.167571   \n",
       "630       GC  AC244090.1  0.112713  0.001160   0.048682         0.128027   \n",
       "631       GC   EPHA1-AS1  0.177610  0.001162   0.048682         0.248932   \n",
       "632       GC       MED31  0.178074  0.001162   0.048682         0.228414   \n",
       "635       GC       KALRN  0.253556  0.001191   0.049684         0.403670   \n",
       "\n",
       "     AveExpr_rest  percentExpr_cluster  percentExpr_rest  \n",
       "10       0.758216             0.882597          0.594000  \n",
       "14       0.053708             0.510821          0.074667  \n",
       "21       1.231422             0.743848          0.732667  \n",
       "24       0.083226             0.454788          0.120000  \n",
       "26       0.201183             0.640972          0.251667  \n",
       "..            ...                  ...               ...  \n",
       "626      0.026268             0.105263          0.030468  \n",
       "630      0.033475             0.105263          0.067737  \n",
       "631      0.052469             0.157895          0.066921  \n",
       "632      0.081405             0.157895          0.143090  \n",
       "635      0.109891             0.263158          0.129489  \n",
       "\n",
       "[13019 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joint_DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>TPM1</td>\n",
       "      <td>1.456768</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.017393</td>\n",
       "      <td>0.758216</td>\n",
       "      <td>0.882597</td>\n",
       "      <td>0.594000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>FXYD5</td>\n",
       "      <td>0.655155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.518204</td>\n",
       "      <td>0.053708</td>\n",
       "      <td>0.510821</td>\n",
       "      <td>0.074667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>LDHA</td>\n",
       "      <td>1.148616</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.727698</td>\n",
       "      <td>1.231422</td>\n",
       "      <td>0.743848</td>\n",
       "      <td>0.732667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>FABP5</td>\n",
       "      <td>0.824986</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.645517</td>\n",
       "      <td>0.083226</td>\n",
       "      <td>0.454788</td>\n",
       "      <td>0.120000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>SLC16A3</td>\n",
       "      <td>0.864479</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.897528</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.640972</td>\n",
       "      <td>0.251667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>PGR</td>\n",
       "      <td>0.135100</td>\n",
       "      <td>0.001135</td>\n",
       "      <td>0.048031</td>\n",
       "      <td>0.167571</td>\n",
       "      <td>0.026268</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.030468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>AC244090.1</td>\n",
       "      <td>0.112713</td>\n",
       "      <td>0.001160</td>\n",
       "      <td>0.048682</td>\n",
       "      <td>0.128027</td>\n",
       "      <td>0.033475</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.067737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>EPHA1-AS1</td>\n",
       "      <td>0.177610</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.048682</td>\n",
       "      <td>0.248932</td>\n",
       "      <td>0.052469</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.066921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>MED31</td>\n",
       "      <td>0.178074</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.048682</td>\n",
       "      <td>0.228414</td>\n",
       "      <td>0.081405</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.143090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>KALRN</td>\n",
       "      <td>0.253556</td>\n",
       "      <td>0.001191</td>\n",
       "      <td>0.049684</td>\n",
       "      <td>0.403670</td>\n",
       "      <td>0.109891</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.129489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13019 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene     logFC   P.Value  adj.P.Val  AveExpr_cluster  \\\n",
       "cluster                                                               \n",
       "VCT_CCC        TPM1  1.456768  0.000000   0.000000         2.017393   \n",
       "VCT_CCC       FXYD5  0.655155  0.000000   0.000000         0.518204   \n",
       "VCT_CCC        LDHA  1.148616  0.000000   0.000000         1.727698   \n",
       "VCT_CCC       FABP5  0.824986  0.000000   0.000000         0.645517   \n",
       "VCT_CCC     SLC16A3  0.864479  0.000000   0.000000         0.897528   \n",
       "...             ...       ...       ...        ...              ...   \n",
       "GC              PGR  0.135100  0.001135   0.048031         0.167571   \n",
       "GC       AC244090.1  0.112713  0.001160   0.048682         0.128027   \n",
       "GC        EPHA1-AS1  0.177610  0.001162   0.048682         0.248932   \n",
       "GC            MED31  0.178074  0.001162   0.048682         0.228414   \n",
       "GC            KALRN  0.253556  0.001191   0.049684         0.403670   \n",
       "\n",
       "         AveExpr_rest  percentExpr_cluster  percentExpr_rest  \n",
       "cluster                                                       \n",
       "VCT_CCC      0.758216             0.882597          0.594000  \n",
       "VCT_CCC      0.053708             0.510821          0.074667  \n",
       "VCT_CCC      1.231422             0.743848          0.732667  \n",
       "VCT_CCC      0.083226             0.454788          0.120000  \n",
       "VCT_CCC      0.201183             0.640972          0.251667  \n",
       "...               ...                  ...               ...  \n",
       "GC           0.026268             0.105263          0.030468  \n",
       "GC           0.033475             0.105263          0.067737  \n",
       "GC           0.052469             0.157895          0.066921  \n",
       "GC           0.081405             0.157895          0.143090  \n",
       "GC           0.109891             0.263158          0.129489  \n",
       "\n",
       "[13019 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joint_DE_table.set_index('cluster', inplace=True)\n",
    "joint_DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "joint_DE_table.to_csv('/lustre/scratch117/cellgen/team292/aa22/with_Luz/202111_MFI_CellPhone/DEGs_list_inv_trophoblast_for_cellphone_20211202.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
